---
output: html_document
title: "Interaction effects"
---

```{r setup, include=F}
knitr::opts_chunk$set(echo = FALSE, message = FALSE,
                      fig.path = "../output/interactions/")
suppressMessages(silent <- lapply(
  c("knitr", "kableExtra", "tidyverse", "cowplot", "patchwork"),
  library, character.only = TRUE
))
```

```{r read-sumstats}
phenos <- read_csv("../data/processed/ukb_phenos_unrelated.csv")

gwis_res_df <- expand_grid(
  e = c("dn3fa", "oily_fish", "fish_oil"),
  y = c("Omega_3_pct", "hscrp_log")
) %>%
  bind_rows(tibble(e = "Omega_3_pct", y = "hscrp_log")) %>%
  rowwise() %>%
  mutate(nom_res = list(read_tsv(paste0("../data/processed/gwis/", e, "_", y, "_merged_nom"))))

# fish_crp_gwis <- read_tsv("../data/processed/gwis/oily_fish_hscrp_log_merged_nom")
# fish_n3_gwis <- read_tsv("../data/processed/gwis/oily_fish_Omega_3_pct_merged_nom")
# n3_crp_gwis <- read_tsv("../data/processed/gwis/Omega_3_pct_hscrp_log_merged_nom")
# fish_oil_crp_gwis <- read_tsv("../data/processed/gwis/fish_oil_hscrp_log_merged_nom")
# fish_oil_n3_gwis <- read_tsv("../data/processed/gwis/fish_oil_Omega_3_pct_merged_nom")

gene_res_df <- expand_grid(
  e = c("dn3fa", "oily_fish", "fish_oil"),
  y = c("Omega_3_pct", "hscrp_log")
) %>%
  bind_rows(tibble(e = "Omega_3_pct", y = "hscrp_log")) %>%
  rowwise() %>%
  mutate(res = list(read_tsv(paste0("../data/processed/magma/", e, "_", y, ".genes.out"))))

# fish_crp_genes <- read_table("../data/processed/magma/oily_fish_hscrp_log.genes.out") %>%
#   arrange(P)
# fish_n3_genes <- read_table("../data/processed/magma/oily_fish_Omega_3_pct.genes.out") %>%
#   arrange(P)
# n3_crp_genes <- read_table("../data/processed/magma/Omega_3_pct_hscrp_log.genes.out") %>%
#   arrange(P)
# fish_oil_crp_genes <- read_table("../data/processed/magma/fish_oil_hscrp_log.genes.out") %>%
#   arrange(P)
# fish_oil_n3_genes <- read_table("../data/processed/magma/fish_oil_Omega_3_pct.genes.out") %>%
#   arrange(P)
```

The goal of this analysis was to understand the genetic modification of the relationship between dietary omega-3 intake and hsCRP, with plasma omega-3 (as a % of FA from NMR) as an established mediator. We conducted a series of genome-wide interaction studies to explore this possibility, with some of these models used to decompose the N3-hsCRP relationship into components "upstream" and "downstream" of the plasma omega-3 mediator.

# GWIS results for the primary dN3FA - hsCRP relationship

```{r prep-qqs}
calc_lambda <- function(x, p=0.5){
  # Calculate genomic inflation lambda value
  x <- x[!is.na(x)]
  x.quantile <- quantile(x, p)
  round(qchisq(1 - x.quantile, 1) / qchisq(p, 1), 2)
}

make_qq <- function(data, pval_col, main=""){
  # Make a quantile-quantile plot
  data <- filter(data, data[[pval_col]] > 0)  # In case extremely low p-values are stored as zero

  # Process p-values
  y_vals <- sort(-log10(data[[pval_col]]))
  x_vals <- -log10(rev(ppoints(length(y_vals))))  # ppoints generates a uniform probability distribution
  lam <- calc_lambda(data[[pval_col]])

  # Trim points at higher p-values (credit to RaMWAS package for code snippet)
  levels = as.integer((x_vals - x_vals[1]) / (tail(x_vals, 1) - x_vals[1]) * 2000)
  keep = c(TRUE, diff(levels) != 0)
  levels = as.integer((y_vals - y_vals[1])/(tail(y_vals, 1) - y_vals[1]) * 2000)
  keep = keep | c(TRUE, diff(levels) != 0)
  keep = which(keep)

  rm(data)  # Deal with memory issues
    
  tibble(x=x_vals[keep], y=y_vals[keep]) %>%
    ggplot(aes(x=x, y=y)) + 
    geom_point() +
    geom_abline(slope=1, linetype="dashed") +
    annotate("text", 
#              label=bquote(lambda == .(lam)),  # NOTE: somehow this bquote call results in the warning about "is.na() applied to type language"
             label=paste("lambda == ", lam), parse=TRUE,
             x=-Inf, y=Inf, hjust=-0.5, vjust=3, size=5) +
    labs(x=expression(-log[10](italic(p)) * " (Expected)"),
         y=expression(-log[10](italic(p)) * " (Observed)"),
         title=main) +
    theme_bw() +
    theme(panel.grid=element_blank())
}
```

```{r prep-manhattans}
make_manhattan <- function(data, 
                           pval_col, snp_col = "SNP", chr_col = "CHR", pos_col = "POS", 
                           threshold = 5e-8, ylims = NULL, main = "",
                           add_labels = TRUE) {

  mh_data <- data %>%
    dplyr::rename(all_of(c(SNP = snp_col, CHR = chr_col, POS = pos_col, P = pval_col))) %>%
    filter(!is.na(P)) %>%
    mutate(P = as.numeric(P),
           P = ifelse(P == 0, min(1e-300, min(P[P != 0], na.rm = TRUE)), P),  # Remove P = 0
           nlp = -log10(P))
  
  # Trim points in crowded regions (credit to RaMWAS package for code snippet)
  yfac = as.integer(mh_data$nlp * 100) + 1L
  yorder = sort.list(yfac)
  yfac <- factor(yfac, levels = as.character(seq_len(max(yfac))))
  ygroup <- split(seq_along(yfac), yfac)
  for (i in seq_along(ygroup)) {
    if (length(ygroup[[i]]) > 300) {
      ygroup[[i]] <- sample(ygroup[[i]], size = 300, replace = FALSE)
    }
  }
  keep <- unlist(ygroup, use.names = FALSE)
  
  mh_data <- mh_data %>%
    select(SNP, CHR, POS, nlp) %>%
    dplyr::slice(keep) %>%
    mutate(POS = as.numeric(as.character(POS)),
           CHR = factor(CHR, levels = c(1:22, "X"))) %>%
    arrange(CHR, POS) %>%
    mutate(pos_idx = seq(1, nrow(.)))
  
  suppressWarnings(chr_lengths <- sapply(c(1:22, "X"), function(chr) {
    with(mh_data, max(POS[CHR == chr], na.rm = TRUE))
  }))
  chr_lengths <- ifelse(is.infinite(chr_lengths), 0, chr_lengths)
  chr_start_pos <- cumsum(chr_lengths) - chr_lengths
  
  mh_data <- mh_data %>%
    mutate(x_coord = chr_start_pos[CHR] + POS,
           color = CHR,
    ) %>%
    arrange(as.integer(color), nlp) #%>%
#     distinct(SNP, .keep_all = T)
  
  lims <- mh_data %>%
    group_by(CHR) %>%
    summarise(avg_coord = (min(x_coord) + max(x_coord)) / 2)
  
  newcols <- setNames(
    rep(x = c("#AAAAAA", "#8A8A8A"), length.out = 23),  # Gray/dark gray for alternating chromosomes
    levels(factor(lims$CHR))
  )
  
  mh_plt <- ggplot() +
    geom_point(data = mh_data, 
               aes(x = x_coord, y = nlp, color = factor(color)), 
               size = 0.75, alpha = 1) +
    geom_hline(yintercept = -log10(threshold), linetype = "dashed", color = "black") + 
    scale_x_continuous(breaks = lims$avg_coord[c(1:16, 18, 20, 20, 22, 23)], 
                       labels = c(1:16, 18, 20, 20, 22, "X"), 
                       expand = c(0,0)) +
    scale_y_continuous(name = expression(-log[10](italic(p)))) +
    scale_colour_manual(values = newcols, guide = "none") +
    scale_fill_manual(name = "Color", values = newcols) +
    labs(title = main) +
    theme(axis.title.x = element_blank(),
          axis.text.x = element_text(vjust = -1.5),
          panel.grid.minor = element_blank(),
          panel.grid.major.x = element_blank())
  if (!is.null(ylims)) mh_plt <- mh_plt + coord_cartesian(ylim = ylims)
  if (add_labels) mh_plt <- mh_plt +
    ggrepel::geom_text_repel(data = filter(mh_data, nlp > -log10(threshold)),
                             aes(x = x_coord, y = nlp, label = SNP))  # ** ggrepel: 53 unlabeled data points (too many overlaps). Consider increasing max.overlaps **
  
  mh_plt
}

gene_bonferroni <- 0.05 / nrow(gene_res_df$res[[1]])
```

```{r primary-manhattans}
make_manhattan(pull(filter(gwis_res_df, e == "dn3fa", y == "hscrp_log"), nom_res)[[1]],
               "robust_P_int", threshold = 5e-8) +
  labs(title = "Variant-level: dN3FA - hsCRP")

include_graphics("../data/processed/gwis/qq_plots/dn3fa_hscrp_log_robust_QQ.pdf")

# make_manhattan(fish_crp_gwis, "robust_P_joint", threshold = 5e-8, add_labels = FALSE) +
#   labs(title = "Variant-level joint test: oily fish - hsCRP")

make_manhattan(pull(filter(gene_res_df, e == "dn3fa", y == "hscrp_log"), res)[[1]],
               "P", snp_col = "GENE", pos_col = "START", threshold = gene_bonferroni) +
  labs(title = "Variant-level: dN3FA - hsCRP")
```

# GWIS results using the plasma omega-3 mediator

We can then explore GWIS results informed by the plasma omega-3 mediator. These corresponding to either "upstream" (genetic modifiers of the fish - plasma N3 relationship) or "downstream" (genetic modifiers of the plasma N3 - hsCRP relationship). These models provide both additional biological insights and may increase statistical power through greater precision (i.e., plasma N3 as a better proxy for true dietary omega-3 intake than self report). 

```{r mediated-manhattans}
make_manhattan(pull(filter(gwis_res_df, e == "dn3fa", y == "Omega_3_pct"), nom_res)[[1]],
               "robust_P_int", threshold = 5e-8) +
  labs(title = "Variant-level: dN3FA - pN3FA")
include_graphics("../data/processed/gwis/qq_plots/dn3fa_Omega_3_pct_robust_QQ.pdf")
make_manhattan(pull(filter(gene_res_df, e == "dn3fa", y == "Omega_3_pct"), res)[[1]],
               "P", snp_col = "GENE", pos_col = "START", threshold = gene_bonferroni) +
  labs(title = "Gene-level: dN3FA - pN3FA (\"upstream\"")

make_manhattan(pull(filter(gwis_res_df, e == "Omega_3_pct", y == "hscrp_log"), nom_res)[[1]],
               "robust_P_int", threshold = 5e-8) +
  labs(title = "Variant-level: pN3FA - hsCRP")
include_graphics("../data/processed/gwis/qq_plots/Omega_3_pct_hscrp_log_robust_QQ.pdf")
make_manhattan(pull(filter(gene_res_df, e == "Omega_3_pct", y == "hscrp_log"), res)[[1]],
               "P", snp_col = "GENE", pos_col = "START", threshold = gene_bonferroni) +
  labs(title = "Gene-level: pN3FA - hsCRP (\"downstream\"")
```

# GWIS results using fish oil instead of oily fish intake

Fish oil is an alternative source of dietary omega-3 intake that, while not highly correlated with fish intake, might be expected to show similar genetic architecture of response modification. We will look at primary and "upstream" GWIS results from models replacing oily fish with fish oil. 

```{r fish-and-fish-oil-manhattans}
make_manhattan(pull(filter(gwis_res_df, e == "oily_fish", y == "Omega_3_pct"), nom_res)[[1]],
               "robust_P_int", threshold = 5e-8) +
  labs(title = "Variant-level: oily fish - pN3FA")
include_graphics("../data/processed/gwis/qq_plots/oily_fish_Omega_3_pct_robust_QQ.pdf")
make_manhattan(pull(filter(gene_res_df, e == "oily_fish", y == "Omega_3_pct"), res)[[1]],
               "P", snp_col = "GENE", pos_col = "START", threshold = gene_bonferroni) +
  labs(title = "Gene-level: oily_fish - pN3FA (\"upstream\"")

make_manhattan(pull(filter(gwis_res_df, e == "fish_oil", y == "Omega_3_pct"), nom_res)[[1]],
               "robust_P_int", threshold = 5e-8) +
  labs(title = "Variant-level: fish oil - pN3FA")
include_graphics("../data/processed/gwis/qq_plots/fish_oil_Omega_3_pct_robust_QQ.pdf")
make_manhattan(pull(filter(gene_res_df, e == "fish_oil", y == "Omega_3_pct"), res)[[1]],
               "P", snp_col = "GENE", pos_col = "START", threshold = gene_bonferroni) +
  labs(title = "Gene-level: fish oil - pN3FA (\"upstream\"")
```

Furthermore, a genetic correlation between (G x Fish) and (G x Fish oil) would strengthen the mechanistic basis for these conclusions about interactions. Do we observe this in practice?

```{r genetic-correlations, eval = F}
hscrp_gc <- read_table("../data/processed/ldsc/oily_fish_fish_oil_hscrp_log.log", 
                       skip = 60, n_max = 1)
n3_gc <- read_table("../data/processed/ldsc/oily_fish_fish_oil_Omega_3_pct.log", 
                       skip = 60, n_max = 1)

gc_summary_df <- bind_rows(list(
  hscrp_gc,
  n3_gc
)) %>%
  select(p1, p2, rg, se, p)
gc_summary_df %>%
  mutate(across(c(p1, p2), basename),
         across(c(rg, se, p), ~ signif(., 2)))
```

# Overall summary

```{r gene-summary-plot}
gene_summary_df <- bind_rows(list(
  full = pull(filter(gene_res_df, e == "dn3fa", y == "hscrp_log"), res)[[1]],
  upstream = pull(filter(gene_res_df, e == "dn3fa", y == "Omega_3_pct"), res)[[1]],
  downstream = pull(filter(gene_res_df, e == "Omega_3_pct", y == "hscrp_log"), res)[[1]]
), .id = "pathway") %>%
  mutate(q = p.adjust(P, method = "BH"),
         fdr_sig = q < 0.05)

make_gene_summary_manhattan <- function(data, 
                                        pval_col = "P", snp_col = "SYMBOL", 
                                        chr_col = "CHR", pos_col = "START", 
                                        threshold = gene_bonferroni, 
                                        ylims = NULL, main = "",
                                        add_labels = TRUE) {
  
  mh_data <- data %>%
    dplyr::rename(all_of(c(SNP = snp_col, CHR = chr_col, POS = pos_col, P = pval_col))) %>%
    filter(!is.na(P)) %>%
    mutate(P = as.numeric(P),
           P = ifelse(P == 0, min(1e-300, min(P[P != 0], na.rm = TRUE)), P),  # Remove P = 0
           nlp = -log10(P))
  
  # Trim points in crowded regions (credit to RaMWAS package for code snippet)
  yfac = as.integer(mh_data$nlp * 100) + 1L
  yorder = sort.list(yfac)
  yfac <- factor(yfac, levels = as.character(seq_len(max(yfac))))
  ygroup <- split(seq_along(yfac), yfac)
  for (i in seq_along(ygroup)) {
    if (length(ygroup[[i]]) > 300) {
      ygroup[[i]] <- sample(ygroup[[i]], size = 300, replace = FALSE)
    }
  }
  keep <- unlist(ygroup, use.names = FALSE)
  
  mh_data <- mh_data %>%
    select(SNP, CHR, POS, nlp, pathway, fdr_sig) %>%
    dplyr::slice(keep) %>%
    mutate(POS = as.numeric(as.character(POS)),
           CHR = factor(CHR, levels = c(1:22, "X"))) %>%
    arrange(CHR, POS) %>%
    mutate(pos_idx = seq(1, nrow(.)))
  
  suppressWarnings(chr_lengths <- sapply(c(1:22, "X"), function(chr) {
    with(mh_data, max(POS[CHR == chr], na.rm = TRUE))
  }))
  chr_lengths <- ifelse(is.infinite(chr_lengths), 0, chr_lengths)
  chr_start_pos <- cumsum(chr_lengths) - chr_lengths
  
  mh_data <- mh_data %>%
    mutate(x_coord = chr_start_pos[CHR] + POS,
           color = ifelse(fdr_sig, pathway, CHR),
    )
  
  lims <- mh_data %>%
    group_by(CHR) %>%
    summarise(avg_coord = (min(x_coord) + max(x_coord)) / 2)
  
  newcols <- setNames(
    c(rep(x = c("#AAAAAA", "#8A8A8A"), length.out = 22),
      RColorBrewer::brewer.pal(3, "Dark2")),
    c(levels(factor(lims$CHR)), unique(mh_data$pathway))
  )
  
  mh_plt <- ggplot() +
    geom_point(data = mh_data, 
               aes(x = x_coord, y = nlp, color = factor(color)), 
               size = 0.75, alpha = 1) +
    geom_hline(yintercept = -log10(threshold), linetype = "dashed", color = "gray") + 
    scale_x_continuous(breaks = lims$avg_coord[c(1:16, 18, 20, 20, 22, 23)], 
                       labels = c(1:16, 18, 20, 20, 22, "X"), 
                       expand = c(0,0)) +
    scale_y_continuous(name = expression(-log[10](italic(p)[int]))) +
    scale_colour_manual(values = newcols, breaks = unique(mh_data$pathway),
                        name = "Sub-pathway") +
    scale_fill_manual(name = "Color", values = newcols) +
    labs(title = main) +
    theme(axis.title.x = element_blank(),
          axis.text.x = element_text(vjust = -1.5),
          panel.grid.minor = element_blank(),
          panel.grid.major.x = element_blank())
  if (!is.null(ylims)) mh_plt <- mh_plt + coord_cartesian(ylim = ylims)
  if (add_labels) mh_plt <- mh_plt +
    ggrepel::geom_text_repel(data = filter(mh_data, fdr_sig),
                             aes(x = x_coord, y = nlp, label = SNP))
  
  mh_plt
}

make_gene_summary_manhattan(gene_summary_df)
```

```{r gene-summary-table}
gene_summary_df %>%
  mutate(locus = paste0(str_pad(CHR, 2, pad = "0"), ":", START, "-", STOP)) %>%
  arrange(P) %>%
  filter(fdr_sig) %>%
  select(pathway, SYMBOL, CHR, START, STOP, locus, NSNPS, P, q) %>%
  write_csv("../data/processed/magma/top_genes.csv")
```

# Gene-specific follow-up

```{r gene-followup-functions}
make_double_strat_plot <- function(top_snp, e, y, df, plt_title) {
  plt_df <- df %>%
    select(all_of(c(G = {{top_snp}}, E = {{e}}, Y = {{y}}))) 
  if (length(unique(plt_df$E)) > 10) {
    plt_df$E <- cut(plt_df$E, 
                    quantile(plt_df$E, seq(0, 1, length.out = 5), na.rm = TRUE),
                    include.lowest = TRUE)
  } else {
    plt_df$E <- factor(round(plt_df$E, 2))
  }
  plt_df %>%
    mutate(G = factor(round(G))) %>%
    filter(!is.na(G), !is.na(E), !is.na(Y)) %>%
    group_by(G, E) %>%
    summarise(m = mean(Y),
              se = sd(Y) / sqrt(n())) %>%
    mutate(l95 = m - 1.96 * se,
           u95 = m + 1.96 * se) %>%
    ggplot(aes(x = G, y = m, color = E)) +
    geom_point(position = position_dodge(width = 0.2)) +
    geom_errorbar(aes(ymin = l95, ymax = u95), width = 0.2,
                  position = position_dodge(width = 0.2)) +
    labs(x = "Genotype", y = "Mean of Y (95% CI)", 
         title = plt_title)
}

make_g_strat_plot <- function(top_snp, e, y, df, plt_title) {
  plt_df <- df %>%
    select(all_of(c(G = {{top_snp}}, E = {{e}}, Y = {{y}}))) 
  plt_df %>%
    mutate(G = factor(round(G))) %>%
    filter(!is.na(G), !is.na(E), !is.na(Y),
           E <= quantile(E, 0.999, na.rm = TRUE)) %>%
    group_by(G) %>%
    filter(n() > 5) %>%
    ggplot(aes(x = E, y = Y, color = G)) +
    geom_smooth(method = "lm") +
    labs(x = "Exposure (cut off at 99th percentile)", y = "Y", 
         title = plt_title)
}

run_gxe <- function(g, e, y, 
                    covars = c("age", "sex"), 
                    df = phenos,
                    std = FALSE) {
  if (std) df <- mutate(df, across(all_of(c(e, y)), ~ scale(.)))
  df$g <- df[[g]]
  lm_str <- paste0(y, " ~ g * ", e, " + ", paste0(covars, collapse = " + "))
  lm_summ <- tryCatch({
    lm_fit <- lm(as.formula(lm_str), data = df) 
    lm_fit %>%
      broom::tidy() %>%
      filter(term == paste0("g:", e)) %>%
      mutate(residual_df = lm_fit$df.residual)
  }, error = tibble(NA))
  lm_summ
}

make_int_effect_plot <- function(all_maf_ss_df, e, top_snpid) {
  beta_vec <- all_maf_ss_df[[paste0("Beta_G-", e)]]
  ylims <- median(beta_vec, na.rm = TRUE) + c(-5, 5) * mad(beta_vec, na.rm = TRUE)
  all_maf_ss_df %>%
    mutate(int_estimate = !!sym(paste0("Beta_G-", e)),
           int_se = !!sym(paste0("SE_Beta_G-", e))) %>%
    arrange(int_estimate) %>%
    mutate(l95 = int_estimate - 1.96 * int_se,
           u95 = int_estimate + 1.96 * int_se,
           top_gwis = (SNPID == gsub("_.*", "", top_snpid)),
           SNPID = factor(SNPID, levels = SNPID)) %>%
    ggplot(aes(x = SNPID, y = int_estimate)) +
    geom_point(aes(color = top_gwis)) +
    geom_errorbar(aes(ymin=l95, ymax=u95, color = top_gwis), width = 0.2) +
    geom_hline(yintercept = 0, color = "gray") +
    # geom_text(aes(label = SNPID)) +
    scale_color_manual(values = c(`FALSE` = "black", `TRUE` = "red"), 
                       breaks = c(TRUE),
                       labels = c("Top variant from GWIS")) +
    labs(x = "SNPID", y = "Interaction effect estimate (95% CI)") +
    # coord_cartesian(ylim = ylims) +
    theme(axis.text.x = element_blank(),
          # axis.text.x = element_text(angle = 45, hjust = 0.9),
          legend.position = "bottom", legend.title = element_blank(), 
          legend.key = element_blank())
}

make_int_effect_maf_plot <- function(all_maf_ss_df, e, top_snpid) {
  beta_vec <- all_maf_ss_df[[paste0("Beta_G-", e)]]
  ylims <- median(beta_vec, na.rm = TRUE) + c(-5, 5) * mad(beta_vec, na.rm = TRUE)
  all_maf_ss_df %>%
    filter(SE_Beta_G < (100 * median(SE_Beta_G))) %>%
    mutate(int_estimate = !!sym(paste0("Beta_G-", e)),
           int_se = !!sym(paste0("SE_Beta_G-", e))) %>%
    mutate(l95 = int_estimate - 1.96 * int_se,
           u95 = int_estimate + 1.96 * int_se,
           top_gwis = (SNPID == gsub("_.*", "", top_snpid)),
           MAF = pmin(AF, 1 - AF)) %>%
    # arrange(top_gwis) %>%
    arrange(annot != "intronic") %>%
    ggplot(aes(x = MAF, y = int_estimate, color = annot)) +
    geom_point() +
    # geom_point(aes(color = top_gwis)) +
    geom_errorbar(aes(ymin=l95, ymax=u95), width = 0, color = "black", alpha = 0.5) +
    geom_hline(yintercept = 0, color="gray") +
    scale_x_log10() +
    # scale_color_manual(values = c(`FALSE` = "black", `TRUE` = "red"), 
    #                    breaks = c(TRUE),
    #                    labels = c("Top variant from GWIS")) +
    labs(x = "MAF", y = "Interaction effect estimate (95% CI)") +
    # coord_cartesian(ylim = ylims) +
    theme(axis.text.x = element_text(angle = 45, hjust = 0.9),
          legend.position = "bottom", legend.title = element_blank(), 
          legend.key = element_blank())
}

make_int_effect_cond_plot <- function(ss_df, ss_cond_df, e, top_snpid) {
  beta_vec <- ss_cond_df[[paste0("Beta_G-", e)]]
  ylims <- median(beta_vec, na.rm = TRUE) + c(-5, 5) * mad(beta_vec, na.rm = TRUE)
  bind_rows(list(unconditional = ss_df, conditional = ss_cond_df), 
            .id = "adjustment") %>%
    filter(SE_Beta_G < (100 * median(SE_Beta_G))) %>%
    group_by(SNPID) %>%
    filter(any(P_Value_Interaction < 0.05)) %>%
    ungroup() %>%
    mutate(int_estimate = !!sym(paste0("Beta_G-", e)),
           int_se = !!sym(paste0("SE_Beta_G-", e))) %>%
    select(SNPID, AF, adjustment, int_estimate, int_se) %>%
    arrange(int_estimate) %>%
    mutate(l95 = int_estimate - 1.96 * int_se,
           u95 = int_estimate + 1.96 * int_se,
           top_gwis = (SNPID == gsub("_.*", "", top_snpid)),
           MAF = pmin(AF, 1 - AF),
           SNPID = factor(SNPID, levels = unique(SNPID[adjustment == "unconditional"]))) %>%
    filter(!top_gwis) %>%
    ggplot(aes(x = SNPID, y = int_estimate, color = adjustment)) +
    geom_point(position = position_dodge(width = 0.1)) +
    geom_errorbar(aes(ymin = l95, ymax = u95), width = 0, 
                  position = position_dodge(width = 0.1)) +
    geom_hline(yintercept = 0, color = "gray") + 
    labs(x = "", 
         y = "Interaction effect (95% CI)") +
    # coord_cartesian(ylim = ylims) +
    theme(axis.text.x = element_text(angle = 45, hjust = 0.9),
          legend.position = "bottom", legend.title = element_blank(), 
          legend.key = element_blank())
}

make_int_p_cond_plot <- function(ss_df, ss_cond_df, e, top_snpid) {
  bind_rows(list(unconditional = ss_df, conditional = ss_cond_df), 
            .id = "adjustment") %>%
    filter(SE_Beta_G < (100 * median(SE_Beta_G))) %>%
    mutate(int_nlp = -log10(P_Value_Interaction)) %>%
    select(SNPID, SNPID, AF, adjustment, int_nlp) %>%
    pivot_wider(names_from = "adjustment", values_from = "int_nlp") %>%
    mutate(top_gwis = (SNPID == gsub("_.*", "", top_snpid)),
           MAF = pmin(AF, 1 - AF)) %>%
    filter(!top_gwis) %>%
    ggplot(aes(x = unconditional, y = conditional, color = MAF)) +
    geom_point() +
    geom_abline(slope = 1, intercept = 0, color="gray") +
    scale_color_continuous(trans = "log", breaks = c(0.001, 0.01, 0.1)) +
    labs(x = "Interaction p-value (unconditional)", 
         y = "Interaction p-value (conditional on top variant)") +
    theme(axis.text.x = element_text(angle = 45, hjust = 0.9),
          legend.position = "bottom")
}

run_gene_follow_up <- function(gene, e, y, 
                               base_covars = covar_sets$ffqAdj_minusAC,
                               snp_idx = 1) {
  
  print(paste0("Plots for gene: ", gene, "(E = ", e, ", Y = ", y, ")"))
  
  e_regression <- if (e == "fish_oil") "fish_oil_touchscreen" else e
  
  # gwis_sumstats_df <- read_tsv(paste0("../data/processed/gene_followup/", 
  #                                     e, "_", y, "_", gene, "_subset"))
  
  top_snp_df <- read_tsv(paste0("../data/processed/gene_followup/", 
                                 e, "_", y, "_", gene, "_top_snp.raw")) %>%
    select(id = IID, last_col()) %>%
    mutate(id = as.character(id))
  top_snp <- names(top_snp_df)[snp_idx + 1]
  regression_df <- inner_join(phenos, top_snp_df, by = "id")
  
  double_strat_plt <- make_double_strat_plot(top_snp, e_regression, y, regression_df,
                                             "Stratified interaction plot for top SNP")
  g_strat_plt <- make_g_strat_plot(top_snp, e_regression, y, regression_df,
                                   "Stratified interaction plot for top SNP")
  print(double_strat_plt + g_strat_plt)
  
  sensitivity_covar_list <- list(
    primary = base_covars,
    add_E_by_gPC = c(base_covars, paste0(e_regression, " * gPC", 1:10)),
    add_G_by_Cov = c(base_covars, paste0("g * ", base_covars)),
    add_BMI = c(base_covars, "bmi")
  )
  sensitivity_covar_res_df <- tibble(covar_set = names(sensitivity_covar_list)) %>%
    rowwise() %>%
    # mutate(model_res = list(run_gxe(top_snp, e_regression, y,
    #                                 sensitivity_covar_list[[covar_set]], 
    #                                 regression_df, std = TRUE))) %>%
    mutate(model_res = list(run_gxe(
      top_snp, e_regression, y,
      sensitivity_covar_list[[covar_set]], 
      regression_df, std = TRUE
    ))) %>%
    unnest(model_res) %>%
    mutate(across(estimate:p.value, ~ signif(., 3)))
  print(sensitivity_covar_res_df)
  
  sensitivity_e_list <- list(
    primary = e_regression,
    oily_fish = "oily_fish",
    nonoily_fish = "nonoily_fish",
    N3FA = "N3FA",
    fish_oil_touchscreen = "fish_oil_touchscreen"
  )
  sensitivity_e_res_df <- tibble(e_test = names(sensitivity_e_list)) %>%
    rowwise() %>%
    mutate(model_res = list(run_gxe(top_snp, sensitivity_e_list[[e_test]], y,
                                    base_covars, 
                                    regression_df, std = TRUE))) %>%
    unnest(model_res) %>%
    mutate(across(estimate:p.value, ~ signif(., 3)))
  print(sensitivity_e_res_df)
  
  # all_maf_sumstats_df <- read_tsv(paste0("../data/processed/gene_followup/", 
  #                                        e, "_", y, "_", gene, "_regressions"))
  # 
  # all_maf_sumstats_df %>%
  #   mutate(CHR = gsub("^chr", "", CHR)) %>%
  #   select(CHR, start = POS, end= POS, Non_Effect_Allele, Effect_Allele, SNPID) %>%
  #   write_tsv(paste0("../data/processed/annovar/", gene, "_annovar_input.tsv"))
  # system(paste0("post_gwis/run_annovar.sh ", gene))
  # variant_annot_df <- read_tsv(
  #   paste0("../data/processed/annovar/", gene, "_variants.variant_function"),
  #   col_names = c("annot", "gene", "chr", "start", "end", "REF", "ALT", "SNPID")
  # ) %>%
  #   select(SNPID, annot)
  # all_maf_sumstats_df <- all_maf_sumstats_df %>%
  #   left_join(variant_annot_df, by = "SNPID") %>%
  #   mutate(annot = ifelse(is.na(annot), "unannotated", annot))
  # 
  # all_maf_cond_sumstats_df <- read_tsv(paste0("../data/processed/gene_followup/", 
  #                                             e, "_", y, "_", gene, "_regressions_cond")) %>%
  #   left_join(variant_annot_df, by = "SNPID") %>%
  #   mutate(annot = ifelse(is.na(annot), "unannotated", annot))
  # 
  # int_effect_maf_plt <- make_int_effect_maf_plot(all_maf_sumstats_df, e, top_snp)
  # # int_effect_cond_maf_plt <- make_int_effect_maf_plot(
  # #   filter(all_maf_cond_sumstats_df, 
  # #          POS != str_remove_all(top_snp, "^chr[0-9]+:|:[ATGC].*$")), 
  # #   e, top_snp
  # # )
  # int_effect_cond_plt <- make_int_effect_cond_plot(all_maf_sumstats_df,
  #                                                  all_maf_cond_sumstats_df,
  #                                                  e, top_snp)
  # int_p_cond_plt <- make_int_p_cond_plot(all_maf_sumstats_df, 
  #                                        all_maf_cond_sumstats_df,
  #                                        e, top_snp)

  # print(int_effect_maf_plt)
  # print(int_effect_cond_plt)
  # print(int_p_cond_plt)
  
  # beta_int_col <- paste0("Beta_G-", e_regression)
  # top_cond_sumstats_df <- all_maf_cond_sumstats_df %>%
  #   filter(SE_Beta_G < (100 * median(SE_Beta_G))) %>%
  #   mutate(MAF = pmin(AF, 1 - AF)) %>%
  #   # arrange(desc(abs(!! sym(beta_int_col)))) %>%
  #   arrange(P_Value_Interaction) %>%
  #   select(SNPID, 
  #          # CHR, POS, Effect_Allele, 
  #          MAF, annot,
  #          {{beta_int_col}}, 
  #          `P-int` = P_Value_Interaction, `P-marg` = P_Value_Marginal) %>%
  #   head(10)
  # print("Top variants in conditional analysis")
  # print(top_cond_sumstats_df)
  
  # top_cond_snp_df <- read_tsv(paste0("../data/processed/gene_followup/", 
  #                                    e, "_", y, "_", gene, "_top_cond_snp.raw")) %>%
  #   select(id = IID, last_col()) %>%
  #   mutate(id = as.character(id))
  # top_cond_snp <- names(top_cond_snp_df)[snp_idx + 1]
  # regression_df <- inner_join(phenos, top_cond_snp_df, by = "id")
  # cond_double_strat_plt <- make_double_strat_plot(top_cond_snp, e_regression, y, regression_df,
  #                                                 "Stratified interaction plot for top conditional rare SNP")
  # cond_g_strat_plt <- make_g_strat_plot(top_cond_snp, e_regression, y, regression_df,
  #                                       "Stratified interaction plot for top conditional rare SNP")
  # print(cond_double_strat_plt + cond_g_strat_plt)
}
```

```{r prep-gene-followup}
phenos <- readRDS("../data/processed/manuscript/phenos.rds") %>%  # Already covariate-imputed
  mutate(id = as.character(id))

minimal_covars <- c("age", "age_squared", "sex", "ageBySex")
ses_hl_covars <- c("ac", "income", "education", "smoking", "alcohol")
ffq_covars <- c("cooked_veg", "raw_veg", "fresh_fruit", "prmeat", "whole_bread")
covar_sets <- list(
  minimal = minimal_covars,
  adj = c(minimal_covars, ses_hl_covars),
  mdsAdj = c(minimal_covars, ses_hl_covars, "mds"),
  ffqAdj = c(minimal_covars, ses_hl_covars, ffq_covars),
  ffqAdj_minusAC = setdiff(c(minimal_covars, ses_hl_covars, ffq_covars), "ac")
)
```

## *FADS* cluster (FADS1, FADS2) (oily fish -> N3 %)

```{r fads-cluster-sumstats, eval=F}
run_gene_follow_up("fads1", "oily_fish", "Omega_3_pct")
run_gene_follow_up("fads2", "oily_fish", "Omega_3_pct")
run_gene_follow_up("tmem258", "oily_fish", "Omega_3_pct")
run_gene_follow_up("fen1", "oily_fish", "Omega_3_pct")
run_gene_follow_up("myrf", "oily_fish", "Omega_3_pct")
run_gene_follow_up("fen1", "oily_fish", "Omega_3_pct")

# gene <- "fads1"
# e <- "oily_fish"
# y <- "Omega_3_pct"
# base_covars <- covar_sets$ffqAdj_minusAC
# snp_idx <- 1
# top_rsid_df <- read_tsv(paste0("../data/processed/gene_followup/", gene, 
#                                "_top_rsid.raw")) %>%
#   select(id = IID, matches("^rs")) %>%
#   mutate(id = as.character(id))
# second_rsid_df <- read_tsv("fads1_second_variant.raw") %>%
#   select(id = IID, matches("^rs")) %>%
#   mutate(id = as.character(id))
# regression_df <- phenos %>%
#   inner_join(top_rsid_df, by = "id") %>%
#   inner_join(second_rsid_df, by = "id") %>%
#   mutate(gxe_covar = oily_fish * rs174550_T)
```

## Additional "upstream" genes (oily fish -> N3 %)

```{r upstream-genes, eval=F}
run_gene_follow_up("sfrp1", "oily_fish", "Omega_3_pct")
# run_gene_follow_up("mmp16", "oily_fish", "Omega_3_pct")
run_gene_follow_up("gltscr1", "oily_fish", "Omega_3_pct")
# run_gene_follow_up("ros1", "oily_fish", "Omega_3_pct")
```

## "Downstream" genes (N3 % -> hsCRP)

```{r downstream-genes, eval=F}
run_gene_follow_up("xrn1", "Omega_3_pct", "hscrp_log")
run_gene_follow_up("mica", "Omega_3_pct", "hscrp_log")
run_gene_follow_up("cbll1", "Omega_3_pct", "hscrp_log")
run_gene_follow_up("slc26a4", "Omega_3_pct", "hscrp_log")
```

## rs149473 (tentatively assigned to *LINC00170*)

```{r rs149473-sumstats, eval=F}
run_gene_follow_up("parp8", "oily_fish", "Omega_3_pct", snp_idx = 2)
```

# Replication

## Power calculations for replication

```{r power-calcs}
library(ESPRESSO.GxE)

param_grid <- expand_grid(
  maf = c(0.1, 0.25),
  gxe_beta = seq(0.01, 0.04, by = 0.005),
  N = c(1500, 25000),
  env_reliability = 1
)

simulation_params <- tibble(
  scenario.id = 1:nrow(param_grid),
  seed.val = 1,
  numsims = 100,
  numcases = 1000,  # These are just to prevent errors - shouldn't affect results
  numcontrols = 1000,  #
  numsubjects = param_grid$N,
  interaction.OR = 1,  #
  interaction.efkt = param_grid$gxe_beta,
  p.val = 0.05,
  power = 0.8
)

pheno_params <- tibble(
  scenario.id = 1:nrow(param_grid),
  pheno.model = 1,
  disease.prev = 0.1,
  pheno.mean = 0, 
  pheno.sd = 1,
  pheno.sensitivity = 1,   # These are just to prevent errors - shouldn't affect results
  pheno.specificity = 1,   #
  pheno.reliability = 1
)

# Genotype parameters
geno_params <- tibble(
  scenario.id = 1:nrow(param_grid),
  geno.model = 1,
  MAF = param_grid$maf,
  geno.efkt = 0.01,
  geno.sensitivity = 1,
  geno.specificity = 1
)

# Environment parameters
env_params <- data.frame(
  scenario.id = 1:nrow(param_grid),
  env.model = 1,
  env.prevalence = 0.1,
  env.efkt = 0.2,
  env.mean = 0,
  env.sd = 1,
  env.low.lim = 0,
  env.up.lim = 1,
  env.reliability = param_grid$env_reliability
)

run_simulations <- function(sp, pp, gp, ep, s2r) {
    system("rm -f output.csv")  # Just to be sure (since ESPRESSO.GxE doesn't overwrite)
    capture.output(run.espresso.GxE(sp, pp, gp, ep, s2r))
    results <- suppressWarnings(read_delim("output.csv", delim = ";"))
    new_colnames <- c(names(results)[1:34], 1:4, names(results)[35:38])
    results_fmt <- read_delim("output.csv", delim = ";", 
                              skip = 1, col_names = FALSE) %>%
      setNames(new_colnames) %>%
      select(numsubjects, interaction.efkt, MAF, env.reliability,
             empirical.power)
    system("rm output.csv")
    results_fmt
}

sim_res <- run_simulations(simulation_params, pheno_params, geno_params, 
                           env_params, 1:nrow(param_grid))

sim_res %>%
  mutate(MAF = factor(MAF)) %>%
  ggplot(aes(x = interaction.efkt, y = empirical.power, 
             group = MAF, color = MAF)) +
  geom_point() +
  geom_line() +
  geom_vline(xintercept = 0.025, linetype = "dotted", color = "black") +
  geom_vline(xintercept = 0.035, linetype = "dotted", color = "black") +
  facet_wrap(vars(numsubjects), nrow = 1, 
             labeller = labeller(numsubjects = c("1500" = "N = 1500 (w/ LC-MS)", 
                                                 "25000"="N = 25000 (whole cohort)"))) +
  labs(x = expression("Interaction effect (" * SD[y] * " / " * SD[e] * " / allele)"), 
       y = expression("Empirical power (" * italic(p) * " < 0.05)"))
```

## Replication

```{r replication-exports}
fetch_top_rsid <- function(gene) {
  top_snp_fn <- list.files(path = "../data/processed/gene_followup/", 
                           pattern = paste0(".*", tolower(gene), "_top_snp.txt"),
                           full.names = TRUE)
  top_snp <- scan(top_snp_fn, what = character(), quiet = TRUE)
  top_snp_pos <- as.numeric(str_match(top_snp, ":([0-9]*):")[2])
  gwis_subset_fn <- list.files(path = "../data/processed/gene_followup/", 
                               pattern = paste0(".*", tolower(gene), "_subset"),
                               full.names = TRUE)
  gwis_subset_df <- read_tsv(gwis_subset_fn)
  top_rsid <- gwis_subset_df$RSID[gwis_subset_df$POS == top_snp_pos]
  if (length(top_rsid) == 1) top_rsid else gsub(":[A-Z]*:[A-Z]*", "", top_snp)
}

# fetch_top_rsid2 <- function(gene) {
#   f2 <- list.files(path = "../data/processed/gene_followup/", 
#                    pattern = paste0(".*", tolower(gene), "_top_rsid2.txt"),
#                   full.names = TRUE)
#   tryCatch(scan(f2, what = character(), quiet = TRUE), error = function(e) "")
# }

replication_export_tbl <- gene_summary_df %>%
  filter(fdr_sig) %>%
  rowwise() %>%
  mutate(primary_rsid = fetch_top_rsid(SYMBOL)) %>%
         # secondary_rsid = fetch_top_rsid2(SYMBOL)) %>%
  ungroup() %>%
  select(pathway, SYMBOL, CHR, primary_rsid) %>%  #, secondary_rsid) %>%
  pivot_longer(contains("_rsid"), 
               names_to = "rsid_type", names_pattern = "(.*)_rsid",
               values_to = "rsid")
write_csv(replication_export_tbl, "../data/processed/wghs/table_for_replication.csv")
write(replication_export_tbl$rsid, "../data/processed/wghs/rsids_for_replication.csv")
```

```{r replication-results}
load("../data/processed/wghs/20231024_export/replication_results.RData")

colnames(pheno_geno_summary) <- str_trim(colnames(pheno_geno_summary))
wghs_rep_af_df <- tibble(rsid = grep("^rs", colnames(pheno_geno_summary), value = TRUE)) %>%
  rowwise() %>%
  mutate(m = pheno_geno_summary[4, rsid]) %>%
  ungroup() %>%
  mutate(AF = as.numeric(gsub(".*:", "", m)) / 2) %>%
  select(rsid, af_wghs = AF)

wghs_rep_df <- replication_p_res_df %>%
  select(rsid = g, e, y, beta_wghs = estimate, p_wghs = p.value) %>%
  left_join(wghs_rep_af_df, by = "rsid")

upstream_rep_df <- replication_export_tbl %>%
  filter(rsid_type == "primary", 
         pathway == "upstream") %>%
  left_join(
    select(pull(filter(gwis_res_df, e == "dn3fa", y == "Omega_3_pct"), nom_res)[[1]], 
           SNP, af_ukb = AF, beta_ukb = `Beta_G-dn3fa`),
    by = c("rsid" = "SNP")
  ) %>%
  left_join(wghs_rep_df, by = c("rsid")) %>%
  filter(e == "N3") %>%
  arrange(p_wghs)
make_qq(filter(upstream_rep_df, e == "N3"), "p_wghs")

downstream_rep_df <- replication_export_tbl %>%
  filter(rsid_type == "primary", 
         pathway == "downstream") %>%
  left_join(
    select(pull(filter(gwis_res_df, e == "Omega_3_pct", y == "hscrp_log"), nom_res)[[1]], 
           SNP, af_ukb = AF, beta_ukb = `Beta_G-Omega_3_pct`),
    by = c("rsid" = "SNP")
  ) %>%
  left_join(wghs_rep_df, by = c("rsid")) %>%
  filter(e == "N3") %>%
  arrange(p_wghs)
make_qq(filter(downstream_rep_df, e == "N3"), "p_wghs")

up_down_rep_df <- bind_rows(list(
  upstream = upstream_rep_df,
  downstream = downstream_rep_df
), .id = "pathway")
```

# Export datasets for manuscript

```{r export-manuscript}
# gwis_summary_df <- bind_rows(list(
#   full = pull(filter(gwis_res_df, e == "dn3fa", y == "hscrp_log"), res)[[1]],
#   upstream = pull(filter(gwis_res_df, e == "dn3fa", y == "Omega_3_pct"), res)[[1]],
#   downstream = pull(filter(gwis_res_df, e == "Omega_3_pct", y == "hscrp_log"), res)[[1]]
# ), .id = "pathway") %>%
#   mutate(q = p.adjust(P, method = "BH"),
#          fdr_sig = q < 0.05)
# saveRDS(gwis_summary_df, "../data/processed/manuscript/gwis_summary_df.rds")

saveRDS(gene_summary_df, "../data/processed/manuscript/gene_summary_df.rds")
# saveRDS(gc_summary_df, "../data/processed/manuscript/gc_summary_df.rds")

saveRDS(sim_res, "../data/processed/manuscript/power_sim_res_df.rds")
saveRDS(up_down_rep_df, "../data/processed/manuscript/up_down_rep_df.rds")
```

# Archive

```{r old-gene-followup-viz-functions, eval=F}
make_gene_snp_plot <- function(ss_df, p_col = "robust_P_Value_Interaction") {
  ss_df %>%
    mutate(nlp_int = -log10(!!sym(p_col))) %>%
    ggplot(aes(x = POS, y = nlp_int)) +
    geom_point() +
    geom_text(aes(label = SNPID)) +
    labs(y=expression(-log[10](p[int])))
}

make_int_effect_chr_plot <- function(all_maf_ss_df, e, top_snpid) {
  beta_vec <- all_maf_ss_df[[paste0("Beta_G-", e)]]
  ylims <- median(beta_vec, na.rm = TRUE) + c(-5, 5) * mad(beta_vec, na.rm = TRUE)
  all_maf_ss_df %>%
    mutate(int_estimate = !!sym(paste0("Beta_G-", e)),
           int_se = !!sym(paste0("SE_Beta_G-", e))) %>%
    arrange(int_estimate) %>%
    mutate(l95 = int_estimate - 1.96 * int_se,
           u95 = int_estimate + 1.96 * int_se,
           top_gwis = (SNPID == gsub("_.*", "", top_snpid)),
           SNPID = factor(SNPID, levels = SNPID)) %>%
    ggplot(aes(x = POS, y = int_estimate)) +
    geom_point(aes(color = top_gwis)) +
    geom_errorbar(aes(ymin=l95, ymax=u95, color = top_gwis), width = 0.2) +
    geom_hline(yintercept = 0, color = "gray") +
    # geom_text(aes(label = SNPID)) +
    scale_color_manual(values = c(`FALSE` = "black", `TRUE` = "red"), 
                       breaks = c(TRUE),
                       labels = c("Top variant from GWIS")) +
    labs(x = "SNPID", y = "Interaction effect estimate (95% CI)") +
    # coord_cartesian(ylim = ylims) +
    theme(axis.text.x = element_blank(),
          # axis.text.x = element_text(angle = 45, hjust = 0.9),
          legend.position = "bottom", legend.title = element_blank(), 
          legend.key = element_blank())
}
```

## AHA EPI Lifestyle abstract

```{r aha-results, eval=F}
fads1_var_annot <- read_table("../data/processed/annovar/fads1_variants.variant_function",
                              col_names = c("annot", "gene", "CHR", "POS", "POS2", "REF", "ALT", "RSID"))
fads1_cond_regressions <- read_tsv("../data/processed/gene_followup/oily_fish_Omega_3_pct_fads1_regressions_cond") %>%
  left_join(select(fads1_var_annot, CHR, POS, annot), by = c("CHR", "POS"))
fads1_regressions <- read_tsv("../data/processed/gene_followup/oily_fish_Omega_3_pct_fads1_regressions") %>%
  left_join(select(fads1_var_annot, CHR, POS, annot), by = c("CHR", "POS"))
nonstd_effect <- fads1_cond_regressions %>%
  filter(AF < 0.01,
         annot == "exonic") %>%
  arrange(P_Value_Interaction) %>%
  slice(1) %>%
  select(RSID, AF, `Beta_G-oily_fish`, annot)
std_effect <- nonstd_effect$`Beta_G-oily_fish` / sd(phenos$Omega_3_pct, na.rm = TRUE) * 
  sd(phenos$oily_fish, na.rm = TRUE)
                                                           

fads2_var_annot <- read_table("../data/processed/annovar/fads2_variants.variant_function",
                              col_names = c("annot", "gene", "CHR", "POS", "POS2", "REF", "ALT", "RSID"))
fads2_cond_regressions <- read_tsv("../data/processed/gene_followup/oily_fish_Omega_3_pct_fads2_regressions_cond") %>%
  left_join(select(fads2_var_annot, CHR, POS, annot), by = c("CHR", "POS"))
```

