---
output: html_document
title: "Final outputs for presentations and manuscript"
---

```{r setup, include = F}
knitr::opts_chunk$set(echo = FALSE, message = FALSE,
                      fig.path = "../output/manuscript/",
                      cache.path = "../cache/manuscript/")
suppressMessages(silent <- lapply(
  c("knitr", "kableExtra", "tidyverse", "cowplot", "patchwork",
    "mediation", "parallel"), 
  library, character.only = TRUE))
select <- dplyr::select
theme_set(theme_bw())
```

```{r load-data}
phenos <- readRDS("../data/processed/manuscript/phenos.rds") %>%
  mutate(id = as.character(id))

pheno_corr_df <- readRDS("../data/processed/manuscript/pheno_corr_df.rds")

covar_sets <- readRDS("../data/processed/manuscript/covar_sets_list.rds")
covar_set_labels <- readRDS("../data/processed/manuscript/covar_set_labels.rds")

lm_res_df <- readRDS("../data/processed/manuscript/lm_res_df.rds")

med_obj_list <- readRDS("../data/processed/manuscript/med_obj_list.rds")

power_sim_res <- readRDS("../data/processed/manuscript/power_sim_res_df.rds")
wghs_rep_df <- readRDS("../data/processed/manuscript/up_down_rep_df.rds")
```

# Simulations

```{r run-simulations, cache=1}
run_upstream_sim <- function(pv_gxe_m, pv_m_y, icc_m) {
  g <- rbinom(N, 2, maf)
  e <- rnorm(N, 0, 1)
  beta_gxe_m <- sqrt(pv_gxe_m)
  m <- rnorm(N, beta_gxe_m * scale(g * e), sqrt(1 - pv_gxe_m))
  beta_m_y <- sqrt(pv_m_y)
  y <- rnorm(N, beta_m_y * m, sqrt(1 - pv_m_y))
  
  m_measured <- rnorm(N, sqrt(icc_m) * m, sqrt(1 - icc_m))
  
  std_res <- lm(y ~ g * e) %>%
    broom::tidy() %>%
    filter(term == "g:e")
  med_res <- lm(m_measured ~ g * e) %>%
    broom::tidy() %>%
    filter(term == "g:e")
  bind_rows(list(std = std_res, med = med_res), .id = "approach")
}

run_downstream_sim <- function(pv_e_m, pv_gxm_y, icc_m) {
  g <- rbinom(N, 2, maf)
  e <- rnorm(N, 0, 1)
  beta_e_m <- sqrt(pv_e_m)
  m <- rnorm(N, beta_e_m * e, sqrt(1 - pv_e_m))
  beta_gxm_y <- sqrt(pv_gxm_y)
  y <- rnorm(N, beta_gxm_y * scale(g * m), sqrt(1 - pv_gxm_y))
  
  m_measured <- rnorm(N, sqrt(icc_m) * m, sqrt(1 - icc_m))
  
  std_res <- lm(y ~ g * e) %>%
    broom::tidy() %>%
    filter(term == "g:e")
  med_res <- lm(y ~ g * m_measured) %>%
    broom::tidy() %>%
    filter(term == "g:m_measured")
  bind_rows(list(std = std_res, med = med_res), .id = "approach")
}

test_upstream_scenario <- function(n_sim, p_thresh, ...) {
  mclapply(1:n_sim, \(x) run_upstream_sim(...), mc.cores = 8) %>%
    bind_rows() %>%
    group_by(approach) %>%
    summarize(pwr = sum(p.value < p_thresh) / n())
}

test_downstream_scenario <- function(n_sim, p_thresh, ...) {
  mclapply(1:n_sim, \(x) run_downstream_sim(...), mc.cores = 8) %>%
    bind_rows() %>%
    group_by(approach) %>%
    summarize(pwr = sum(p.value < p_thresh) / n())
}

N <- 1000
maf <- 0.25
n_sim <- 500
alpha <- 0.05

upstream_sim_res_df <- expand_grid(
  pv_gxe_m = seq(0, 0.01, 0.005),
  pv_m_y = seq(0.1, 1, 0.1),
  icc_m = c(0.25, 0.5, 0.75, 1)
) %>%
  rowwise() %>%
  mutate(sim_res = list(test_upstream_scenario(n_sim, alpha, pv_gxe_m, pv_m_y, icc_m))) %>%
  unnest(sim_res) %>%
  mutate(pwr_se = sqrt(pwr * (1 - pwr) / n_sim),
         pwr_l95 = pwr - 1.96 * pwr_se,
         pwr_u95 = pwr + 1.96 * pwr_se)

downstream_sim_res_df <- expand_grid(
  pv_e_m = seq(0.1, 1, 0.1),
  pv_gxm_y = seq(0, 0.01, 0.005),
  icc_m = seq(0.25, 1, 0.25)
) %>%
  rowwise() %>%
  mutate(sim_res = list(test_downstream_scenario(n_sim, alpha, pv_e_m, pv_gxm_y, icc_m))) %>%
  unnest(sim_res) %>%
  mutate(pwr_se = sqrt(pwr * (1 - pwr) / n_sim),
         pwr_l95 = pwr - 1.96 * pwr_se,
         pwr_u95 = pwr + 1.96 * pwr_se)
```

```{r plot-simulations, fig.asp=1.2}
upstream_sim_res_df %>%
  filter(icc_m == 1) %>%
  mutate(across(c(pv_gxe_m), ~ factor(.x)),
         approach = factor(approach,
                           levels = c("std", "med"),
                           labels = c("Standard", "Upstream"))) %>%
  rename(`PV (GxE -> M)` = pv_gxe_m) %>%
  ggplot(aes(x = pv_m_y, y = pwr, group = approach, color = approach)) +
  geom_point() +
  geom_line() +
  scale_color_discrete(name = "Approach") + 
  facet_wrap(vars(`PV (GxE -> M)`), ncol = 1, labeller = label_both) +
  labs(x = "Proportion of variance in Y explained by M",
       y = "Statistical power")

downstream_sim_res_df %>%
  filter(icc_m == 1) %>%
  mutate(across(c(pv_gxm_y), ~ factor(.x)),
         approach = factor(approach,
                           levels = c("std", "med"),
                           labels = c("Standard", "Downstream"))) %>%
  rename(`PV (GxM -> Y)` = pv_gxm_y) %>%
  ggplot(aes(x = pv_e_m, y = pwr, group = approach, color = approach)) +
  geom_point() +
  geom_line() +
  scale_color_discrete(name = "Approach") + 
  facet_wrap(vars(`PV (GxM -> Y)`), ncol = 1, labeller = label_both) +
  labs(x = "Proportion of variance in M explained by E",
       y = "Statistical power")

upstream_example_subset <- filter(upstream_sim_res_df,
                                  pv_gxe_m == 0.005,
                                  pv_m_y == 0.1)
kable(upstream_example_subset, caption = "Upstream simulation example subset")
print(with(upstream_example_subset,
           paste0("Power ratio: ", pwr[approach == "med"] / pwr[approach == "std"])))

downstream_example_subset <- filter(downstream_sim_res_df,
                                    pv_gxm_y == 0.005,
                                    pv_e_m == 0.1)
kable(downstream_example_subset, caption = "Downstream simulation example subset")
print(with(downstream_example_subset,
           paste0("Power ratio: ", pwr[approach == "med"] / pwr[approach == "std"])))
```

```{r plot-simulations-all, fig.asp=1}
all_upstream_sims_plt <- upstream_sim_res_df %>%
  filter(!(approach == "std" & icc_m != 1)) %>%
  mutate(
    label = factor(if_else(approach == "std", "Standard",
                           paste0("Upstream (ICC_M=", icc_m, ")"))),
    label = fct_relevel(label, "Standard", after = 0)
  ) %>%
  mutate(across(c(pv_gxe_m), ~ factor(.x))) %>%
  rename(`PV (GxE -> M)` = pv_gxe_m) %>%
  ggplot(aes(x = pv_m_y, y = pwr, group = label, color = label)) +
  geom_point() +
  geom_line() +
  scale_color_discrete(name = "Approach") +
  facet_wrap(vars(`PV (GxE -> M)`), labeller = label_both) +
  labs(x = "Proportion of variance in Y explained by M",
       y = "Statistical power")

all_downstream_sims_plt <- downstream_sim_res_df %>%
  filter(!(approach == "std" & icc_m != 1)) %>%
  mutate(
    label = factor(if_else(approach == "std", "Standard",
                           paste0("Downstream (ICC_M=", icc_m, ")"))),
    label = fct_relevel(label, "Standard", after = 0)
  ) %>%
  mutate(across(c(pv_gxm_y), ~ factor(.x))) %>%
  rename(`PV (GxM -> Y)` = pv_gxm_y) %>%
  ggplot(aes(x = pv_e_m, y = pwr, group = label, color = label)) +
  geom_point() +
  geom_line() +
  scale_color_discrete(name = "Approach") +
  facet_wrap(vars(`PV (GxM -> Y)`), labeller = label_both) +
  labs(x = "Proportion of variance in M explained by E",
       y = "Statistical power")

all_upstream_sims_plt / all_downstream_sims_plt +
  plot_annotation(tag_levels = "a")
```

Extracting a few key summary data points:

# Basic summaries

```{r table-1}
print("Placeholder for Table 1")
```

```{r dag}
n3fa_hscrp_dag_spec <- dagitty::dagitty(
  'dag {
"hsCRP" [outcome,pos="0.186,-0.622"]
"Healthy lifestyle" [pos="-0.067,-1.093"]
"dN3FA" [exposure,pos="-0.308,-0.615"]
"pN3FA" [exposure,pos="-0.068,-0.615"]
"Socioeconomic status" [pos="-0.069,-1.261"]
Adiposity [pos="-0.068,-0.843"]
"Healthy lifestyle" -> "hsCRP"
"Healthy lifestyle" -> "dN3FA"
"Healthy lifestyle" -> Adiposity
"Socioeconomic status" -> "hsCRP" [pos="0.081,-1.109"]
"Socioeconomic status" -> "Healthy lifestyle" [pos="-0.068,-1.185"]
"Socioeconomic status" -> "dN3FA" [pos="-0.210,-1.096"]
"Socioeconomic status" -> Adiposity [pos="-0.179,-1.044"]
Adiposity -> "hsCRP"
dN3FA -> pN3FA
pN3FA -> hsCRP
}'
)

pdf("../data/processed/manuscript/main_dag.pdf")
plot(n3fa_hscrp_dag_spec)
dev.off()

n3fa_hscrp_dag <- ggdraw() + 
  draw_image(magick::image_read_pdf("../data/processed/manuscript/main_dag.pdf", density = 300))
```

# Main effects analysis

```{r model-selection, fig.asp = 1}
pufa_metabolites <- c("Omega_3", "Omega_6",
                      "Omega_3_pct", "Omega_6_pct",
                      "Omega_3_pct_PUFA", "Omega_6_pct_PUFA",
                      "Omega_6_by_Omega_3",
                      "DHA", "DHA_pct")

n3fa_variables <- c("oily_fish", "nonoily_fish", 
                    paste0("fish_oil", c("_touchscreen", "_verbal")),
                    "FISH", "OILY_FISH",  "fish_oil_24hr",
                    "N3FA", "N6FA")
n3fa_variables_clean <- c(
  "Oily fish - FFQ", "Non-oily fish - FFQ",
  "Fish oil - touchscreen", "Fish oil - verbal interview",
  "Fish - 24HR", "Oily fish - 24HR", "Fish oil - 24HR",
  "Dietary N3FA - 24HR", "Dietary N6FA - 24HR"
)
n3fa_variables <- n3fa_variables[!grepl("24HR", n3fa_variables_clean)]
n3fa_variables_clean <- n3fa_variables_clean[!grepl("24HR", n3fa_variables_clean)]

covar_sets <- covar_sets[!grepl("mds", names(covar_sets))]
covar_set_labels <- covar_set_labels[!grepl("mds", names(covar_set_labels))]

n3fa_var_corr_plt <- pheno_corr_df %>%
  filter(var1 %in% n3fa_variables,
         var2 %in% n3fa_variables) %>%
  mutate(across(c(var1, var2), 
                ~ factor(., levels = n3fa_variables, 
                         labels = n3fa_variables_clean))) %>%
  ggplot(aes(x = var1, y = fct_rev(var2), fill = estimate)) +
  geom_tile() + 
  scale_fill_gradient2(name = "Pearson\ncorrelation") +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(angle = 30, hjust = 0.9))

dv_choice_plt <- lm_res_df %>%
  filter(bm == "hscrp_log",
         dv %in% n3fa_variables,
         !grepl("mds", covar_set)) %>%
  mutate(l95 = estimate - 1.96 * std.error,
         u95 = estimate + 1.96 * std.error,
         dv = factor(dv, levels = n3fa_variables, 
                     labels = n3fa_variables_clean),
         covar_set = factor(covar_set,
                            levels = names(covar_sets),
                            labels = covar_set_labels)) %>%
  ggplot(aes(x = dv, y = estimate, color = covar_set)) +
  geom_point(position = position_dodge(width = 0.4)) +
  geom_errorbar(aes(ymin = l95, ymax = u95),
                position = position_dodge(width = 0.4), width = 0.2) +
  geom_hline(yintercept = 0, color = "gray") + 
  scale_color_discrete(name = "Adjustment set") + 
  labs(x = "Diet variable",
       y = "Standardized effect estimate (95% CI)") +
  theme(axis.text.x = element_text(angle = 30, hjust = 0.9))

# covar_choice_plt <- lm_res_df %>%
#   filter(bm == "hscrp_log",
#          dv == "oily_fish") %>%
#   mutate(l95 = estimate - 1.96 * std.error,
#          u95 = estimate + 1.96 * std.error,
#          covar_set = factor(covar_set, levels = names(covar_sets),
#                             labels = covar_set_labels),
#          dv = factor(dv, levels = n3fa_variables, 
#                      labels = n3fa_variables_clean),
#          bm = factor(bm, levels = c("hscrp_log", "tg_log"),
#                      labels = c("log(hsCRP)", "log(TG)"))) %>%
#   ggplot(aes(x = covar_set, y = estimate)) +
#   geom_point(position = position_dodge(width = 0.4)) +
#   geom_errorbar(aes(ymin = l95, ymax = u95),
#                 position = position_dodge(width = 0.4), width = 0.2) +
#   geom_hline(yintercept = 0, color = "gray") + 
#   # scale_color_discrete(name = "Adjustment set") + 
#   labs(x = "Covariate set",
#        y = "Standardized effect estimate (95% CI)") +
#   facet_wrap(~bm, nrow = 2, scales = "free_y") +
#   theme(axis.text.x = element_text(angle = 30, hjust = 0.9))

(n3fa_hscrp_dag + free(n3fa_var_corr_plt) + plot_layout(widths = c(2, 3))) /
  free(dv_choice_plt) +
  plot_annotation(tag_levels = "a")
```

```{r dn3fa-imputation}
dn3fa_impute_weights_df <- readRDS("../data/processed/manuscript/n3fa_impute_weights_df.rds")

dn3fa_impute_weights_df %>%
  write_csv("../output/manuscript/n3fa_impute_weights_df.csv")

dn3fa_impute_weights_df %>%
  kable(caption = "Weights for fish and fish oil-based N3FA intake estimation:") %>%
  kable_styling(full_width = FALSE)
```

```{r mediator-selection, fig.asp = 0.6}
n3fa_pufa_corr_heatmap_df <- pheno_corr_df %>%
  filter(var1 %in% c("oily_fish", "nonoily_fish", "fish_oil_touchscreen", "fish_oil_verbal"), 
         var2 %in% pufa_metabolites) %>%
  arrange(estimate) %>%
  mutate(var1 = fct_relevel(var1, c("oily_fish", "nonoily_fish", "fish_oil_touchscreen", "fish_oil_verbal")),
         var2 = fct_relevel(var2, unique(var2)),  # For value-based ordering
         var1 = fct_relabel(var1, ~ n3fa_variables_clean[match(., n3fa_variables)])) 
n3fa_pufa_corr_heatmap <- n3fa_pufa_corr_heatmap_df %>%
  ggplot(aes(x = var1, y = var2, fill = estimate)) +
  geom_tile() + 
  scale_fill_gradient2(name = "Pearson\ncorrelation") +
  labs(x = "", y = "") +
  theme(axis.text.x = element_text(
    angle = 30, hjust = 0.9
  ))

dn3fa_impute_weights_plt <- dn3fa_impute_weights_df %>%
  mutate(term = factor(term, 
                       levels = term, 
                       labels = n3fa_variables_clean[match(term, n3fa_variables)])) %>%
  ggplot(aes(x = term, y = weight)) +
  geom_bar(stat = "identity") +
  labs(x = "",
       y = "Regression weight (N3FA% ~ N3FA sources)") +
  theme(axis.text.x = element_text(angle = 30, hjust = 0.9))

n3fa_pufa_corr_heatmap + dn3fa_impute_weights_plt +
  plot_layout(widths = c(1, 2))
```

```{r mediation}
std_med_res_df <- readRDS("../data/processed/manuscript/std_med_res_df.rds") %>%
  mutate(tau.ci.lower = map_dbl(list(med_obj), \(x) x$tau.ci[1]),
         tau.ci.upper = map_dbl(list(med_obj), \(x) x$tau.ci[2]),
         d.ci.lower = map_dbl(list(med_obj), \(x) x$d.avg.ci[1]),
         d.ci.upper = map_dbl(list(med_obj), \(x) x$d.avg.ci[2]),
         z.ci.lower = map_dbl(list(med_obj), \(x) x$z.avg.ci[1]),
         z.ci.upper = map_dbl(list(med_obj), \(x) x$z.avg.ci[2]),
         d.coef = d.avg, z.coef = z.avg)

main_med_plt <- std_med_res_df %>%
  filter(e == "dn3fa", m == "Omega_3_pct", y == "hscrp_log") %>%
  select(contains("coef"), contains("ci")) %>%
  pivot_longer(everything(), names_to = c("estimate", "quantity"), 
               names_pattern = "([a-z]+)\\.(.*)", 
               values_to = "value") %>%
  pivot_wider(names_from = "quantity", values_from = "value") %>%
  mutate(estimate = factor(estimate, levels = c("tau", "d", "z"),
                           labels = c("Total effect", "Mediated effect", "Direct effect"))) %>%
  ggplot(aes(x = estimate, y = coef)) +
  geom_point() +
  geom_errorbar(aes(ymin = ci.lower, ymax = ci.upper), width = 0) +
  geom_hline(yintercept = 0, color = "gray") +
  labs(x = "Mediation quantity", y = "Estimate (dN3FA - pN3FA - hsCRP)") +
  theme(axis.text.x = element_text(angle = 15, hjust = 0.5))

subspecies_med_plt <- std_med_res_df %>%
  filter(e == "dn3fa", y == "hscrp_log") %>%
  select(m, d.coef, contains("d.ci")) %>%
  pivot_longer(-m, names_to = c("estimate", "quantity"), 
               names_pattern = "([a-z]+)\\.(.*)", 
               values_to = "value") %>%
  pivot_wider(names_from = "quantity", values_from = "value") %>%
  mutate(m = factor(m, levels = c("Omega_3_pct", "DHA_pct", "nonDHA_pct"),
                    labels = c("N3FA %", "DHA %", "Non-DHA %"))) %>%
  ggplot(aes(x = m, y = coef)) +
  geom_point() +
  geom_errorbar(aes(ymin = ci.lower, ymax = ci.upper), width = 0) +
  geom_hline(yintercept = 0, color = "gray") +
  labs(x = "Plasma subspecies", y = "ACME estimate (dN3FA - M - hsCRP)")

std_med_res_df %>%
  select(Exposure = e, Mediator = m, Outcome = y,
         `Total effect` = tau.coef, `P - total` = tau.p,
         `Mediated effect` = d.avg, `P - mediated` = d.avg.p,
         `Direct effect` = z.avg, `P - direct` = z.avg.p) %>%
  mutate(across(contains("effect"), ~ round(.x, 3))) %>%
  write_csv("../output/manuscript/mediation_results.csv")
```

```{r neg-control-selection}
nc_justification_plt <- phenos %>%
  mutate(smoking_ever = smoking != "Never",
         college = education == "College or University degree",
         high_income = income %in% c("52,000 to 100,000", "Greater than 100,000")) %>%
  select(oily_fish, raw_veg, hscrp_log, smoking_ever, college, high_income) %>%
  cor(use = "pairwise.complete.obs") %>%
  as_tibble(rownames = "var1") %>%
  pivot_longer(-var1, names_to = "var2", values_to = "corr") %>%
  filter(var1 %in% c("oily_fish", "raw_veg"),
         !(var2 %in% c("oily_fish", "raw_veg"))) %>%
  mutate(var2 = fct_relevel(var2, "hscrp_log")) %>%
  ggplot(aes(x = var2, y = corr, fill = var1)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  scale_fill_discrete(name = "FFQ trait") +
  labs(x = "Variable", y = "Pearson correlation")
```

```{r mediation-negative-control, fig.asp = 0.6}
std_nc_med_res_df <- readRDS("../data/processed/manuscript/std_nc_med_res_df.rds") %>%
  mutate(tau.ci.lower = map_dbl(list(med_obj), \(x) x$tau.ci[1]),
         tau.ci.upper = map_dbl(list(med_obj), \(x) x$tau.ci[2]),
         d.ci.lower = map_dbl(list(med_obj), \(x) x$d.avg.ci[1]),
         d.ci.upper = map_dbl(list(med_obj), \(x) x$d.avg.ci[2]),
         z.ci.lower = map_dbl(list(med_obj), \(x) x$z.avg.ci[1]),
         z.ci.upper = map_dbl(list(med_obj), \(x) x$z.avg.ci[2]),
         d.coef = d.avg, z.coef = z.avg)

nc_med_plt <- std_nc_med_res_df %>%
  select(e, contains("coef"), contains("ci")) %>%
  pivot_longer(-e, names_to = c("estimate", "quantity"),
               names_pattern = "([a-z]+)\\.(.*)",
               values_to = "value") %>%
  pivot_wider(names_from = "quantity", values_from = "value") %>%
  mutate(estimate = factor(estimate, levels = c("tau", "d", "z"),
                           labels = c("Total effect", "Mediated effect", "Direct effect"))) %>%
  ggplot(aes(x = e, y = coef, fill = estimate)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  # geom_errorbar(aes(ymin = ci.lower, ymax = ci.upper), width = 0) +
  geom_hline(yintercept = 0, color = "gray") +
  scale_fill_discrete(name = "") +
  labs(x = "Mediation quantity", y = "Estimate (dN3FA - pN3FA - hsCRP)")

nc_justification_plt / nc_med_plt +
  plot_layout(heights = c(1, 2)) +
  plot_annotation(tag_levels = "a")
```

```{r main-effect-figure, fig.asp = 1}
dn3fa_pn3fa_smooth_plt <- phenos %>%
  ggplot(aes(x = dn3fa, y = Omega_3_pct)) +
  geom_smooth(method = "gam", formula = y ~ rms::rcs(x, 3)) +
  labs(x = "dN3FA", y = "pN3FA")

pn3fa_hscrp_smooth_plt <- phenos %>%
  ggplot(aes(x = Omega_3_pct, y = hscrp)) +
  geom_smooth(method = "gam", formula = y ~ rms::rcs(x, 3)) +
  labs(x = "pN3FA", y = "hsCRP (raw)")

dn3fa_hscrp_smooth_plt <- phenos %>%
  ggplot(aes(x = dn3fa, y = hscrp)) +
  geom_smooth(method = "gam", formula = y ~ rms::rcs(x, 3)) +
  labs(x = "dN3FA", y = "hsCRP (raw)")

smooths <- dn3fa_pn3fa_smooth_plt + pn3fa_hscrp_smooth_plt + dn3fa_hscrp_smooth_plt
smooths / (main_med_plt + subspecies_med_plt) +
  plot_annotation(tag_levels = "a")
```

# Genome-wide summaries

```{r load-gwis-results}
full_gwis_df <- read_tsv("../data/processed/gwis/dn3fa_hscrp_log_merged_nom")
upstream_gwis_df <- read_tsv("../data/processed/gwis/dn3fa_Omega_3_pct_merged_nom")
downstream_gwis_df <- read_tsv("../data/processed/gwis/Omega_3_pct_hscrp_log_merged_nom")
```

```{r gwis-manhattan-variants}
# Make a Miami plot -- credit to RaMWAS package (pruning unneeded points)
# and hudson package (basis for Miami plot functionality)

# data <- filter(data, data[[pval_col_1]] > 0, data[[pval_col_2]] > 0)
nlps1 <- -log10(upstream_gwis_df[["robust_P_int"]])
nlps2 <- -log10(downstream_gwis_df[["robust_P_int"]])

# Trim points in crowded regions (credit to RaMWAS package for code snippet)
yfac = as.integer(nlps1 * 100) + 1L
yorder = sort.list(yfac)
levels(yfac) = as.character(seq_len(max(yfac)))
class(yfac) = "factor"
ygroup = split(seq_along(yfac), yfac)
for (i in seq_along(ygroup)) {
  if (length(ygroup[[i]]) > 300) {
    ygroup[[i]] = sample(ygroup[[i]], size=300, replace=FALSE)
  }
}
keep1 = unlist(ygroup, use.names=FALSE)
yfac = as.integer(nlps2 * 100) + 1L
yorder = sort.list(yfac)
levels(yfac) = as.character(seq_len(max(yfac)))
class(yfac) = "factor"
ygroup = split(seq_along(yfac), yfac)
for (i in seq_along(ygroup)) {
  if (length(ygroup[[i]]) > 300) {
    ygroup[[i]] = sample(ygroup[[i]], size=300, replace=FALSE)
  }
}
keep2 = unlist(ygroup, use.names=FALSE)

top <- cbind(select(upstream_gwis_df, SNP, CHR, POS), pvalue=nlps1)[keep1, ]
bottom=cbind(select(downstream_gwis_df, SNP, CHR, POS), pvalue=nlps2)[keep2, ]

topn <- names(top)
bottomn <- names(bottom)
top$Location <- "Top"
bottom$Location <- "Bottom"
d <- rbind(top, bottom)
d$POS <- as.numeric(as.character(d$POS))
d$CHR <- factor(d$CHR, levels = c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "X", "Y"))
d_order <- d[order(d$CHR, d$POS), ]
d_order$pos_index <- seq.int(nrow(d_order))
chr_lengths <- sapply(1:22, function(chr) max(d[d$CHR == chr, "POS"]))
chr_start_pos <- cumsum(chr_lengths) - chr_lengths
d_order$x_coord <- chr_start_pos[d_order$CHR] + d_order$POS
d_order_sub <- d_order[, c("SNP", "CHR", "POS", "pvalue", "pos_index", "x_coord")]
maxRows <- by(d_order_sub, d_order_sub$CHR, function(x) x[which.max(x$x_coord),])
minRows <- by(d_order_sub, d_order_sub$CHR, function(x) x[which.min(x$x_coord),])
milimits <- do.call(rbind, minRows)
malimits <- do.call(rbind, maxRows)
lims <- merge(milimits, malimits, by="CHR")
names(lims) <- c("Color", 
                 "snpx", "posx", "px", "posidxx", "xcoordmin",
                 "snpy", "posy", "py", "posidxy", "xcoordmax")
lims$av <- (lims$xcoordmin + lims$xcoordmax)/2
lims <- lims[order(lims$Color),]

# gw_int_snps <- whr_ss$RSID[whr_ss$P_Value_Interaction < 5e-8]
# gw_joint_noMarg_snps <- whr_ss$RSID[whr_ss$P_Value_Joint < 5e-8 &
#                                     whr_ss$P_Value_Main > 5e-8]
# d_order$Color <- case_when(
#   d_order$SNP %in% int_marg_snps ~ "30",
#   d_order$SNP %in% int_nomarg_snps ~ "31",
#   d_order$SNP %in% joint_nomarg_snps ~ "32",
#   TRUE ~ as.character(d_order$CHR)
# )
d_order$Color <- as.character(d_order$CHR)
newcols <-c(rep(x=c("#AAAAAA", "#8A8A8A"), length.out=22, each=1),  # Gray/dark gray for alternating chromosomes
            RColorBrewer::brewer.pal(3, "Dark2")[c(2, 3, 1)])
names(newcols) <-c(levels(factor(lims$Color)), "30", "31", "32")
d_order <- arrange(d_order, as.integer(Color), desc(pvalue)) %>%
  distinct(SNP, Location, .keep_all=T)

# TOP PLOT
p1 <- ggplot() +
  geom_point(data=d_order[d_order$Location=="Top", ], 
             aes(x=x_coord, y=pvalue, color=factor(Color)), 
             size=0.75, alpha=1) +
  geom_hline(yintercept=-log10(5e-8), linetype="dashed", color="black") + 
  scale_x_continuous(breaks=lims$av[c(1:16, 18, 20, 20, 22)], 
                     labels=lims$Color[c(1:16, 18, 20, 20, 22)], 
                     expand=c(0,0)) +
  scale_y_continuous(limits=c(0, 15), breaks = seq(0, 15, 5), expand = c(0, 0),
                     name=expression(-log[10](italic(p)[int]) - Upstream)) +
  scale_colour_manual(name = "Color", values = newcols) +
  scale_fill_manual(name = "Color", values = newcols) +
  theme(axis.title.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_text(vjust = -1.5),
        panel.grid.minor = element_blank(),
        panel.grid.major.x = element_blank())

# BOTTOM PLOT
p2 <- ggplot(data=filter(d_order, Location == "Bottom"),
             aes(x=x_coord, y=pvalue, color=factor(Color))) +
  geom_point(size=0.75, alpha=1) +
  geom_hline(yintercept=-log10(5e-8), linetype="dashed", color="black") + 
  scale_x_continuous(breaks=lims$av[c(1:16, 18, 20, 20, 22)], 
                     labels=lims$Color[c(1:16, 18, 20, 20, 22)], 
                     expand=c(0,0)) +
  scale_colour_manual(name = "Color", values = newcols,
                      # breaks=c(30, 31),
                      # labels=c("Genome-wide significant joint test, no genome-wide significant marginal test",
                      #          "Genome-wide significant interaction test")
  ) +
  scale_fill_manual(name = "Color", values = newcols) +
  ylab(expression(-log[10](p) - Interaction)) +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank(),
        panel.grid.minor = element_blank(),
        panel.grid.major.x = element_blank()) +
  scale_y_reverse(limits=c(15,0), breaks = seq(0, 15, 5), expand = c(0, 0),
                  name=expression(-log[10](italic(p)[int]) - Downstream))

# color_legend <- get_legend(
#   p2 + 
#     theme(
#       legend.title=element_blank(), 
#       legend.background=element_rect(size=rel(0.5)),
#       legend.key.size=unit(0.1, "lines"),
#       legend.text=element_text(size=10)
#     ) + 
#     guides(color=guide_legend(override.aes=list(size=2.5)))
# )

(p1 + guides(color = FALSE)) /  
  (p2 + guides(color = FALSE))
```

```{r gmv-chatgpt}
# Calculate -log10(p-values) for upstream and downstream data
nlps1 <- -log10(upstream_gwis_df[["robust_P_int"]])
nlps2 <- -log10(downstream_gwis_df[["robust_P_int"]])

# Function to trim points in crowded regions
trim_points <- function(nlps) {
  yfac <- factor(as.integer(nlps * 100) + 1L)
  ygroup <- split(seq_along(yfac), yfac)
  keep <- unlist(lapply(ygroup, function(x) if (length(x) > 300) sample(x, 300) else x), use.names = FALSE)
  return(keep)
}

# Apply trimming function and select required columns
top <- upstream_gwis_df[trim_points(nlps1), c("SNP", "CHR", "POS")] %>%
  mutate(pvalue = nlps1[trim_points(nlps1)], Location = "Top")
bottom <- downstream_gwis_df[trim_points(nlps2), c("SNP", "CHR", "POS")] %>%
  mutate(pvalue = nlps2[trim_points(nlps2)], Location = "Bottom")

# Combine and process data for plotting
d_order <- rbind(top, bottom) %>%
  mutate(POS = as.numeric(as.character(POS)),
         CHR = factor(CHR, levels = c(as.character(1:22), "X", "Y")),
         Color = as.character(CHR)) %>%
  arrange(CHR, POS) %>%
  mutate(pos_index = seq.int(n()),
         x_coord = ave(POS, CHR, FUN = cumsum))

# Calculate chromosome limits for plotting
chr_limits <- d_order %>%
  group_by(CHR) %>%
  summarize(start_pos = min(x_coord), end_pos = max(x_coord)) %>%
  mutate(mid_pos = (start_pos + end_pos) / 2) %>%
  ungroup()

# Plotting colors
newcols <- c(rep(x = c("#AAAAAA", "#8A8A8A"), length.out = 22, each = 1), RColorBrewer::brewer.pal(3, "Dark2")[c(2, 3, 1)])
names(newcols) <- c(as.character(1:22), "30", "31", "32")

# Base plot with common aesthetics
base_plot <- ggplot() +
  scale_x_continuous(breaks = chr_limits$mid_pos, labels = chr_limits$CHR, expand = c(0, 0)) +
  scale_colour_manual(values = newcols) +
  theme(panel.grid = element_blank(), axis.ticks.x = element_blank())

# Top plot
p1 <- base_plot +
  geom_point(data = filter(d_order, Location == "Top"), aes(x = x_coord, y = pvalue, color = Color), size = 0.75) +
  scale_y_continuous(name = expression(-log[10](italic(p)[int]) - " Upstream"), limits = c(0, 15), breaks = seq(0, 15, 5)) +
  theme(axis.title.x = element_blank(), axis.text.x = element_text(vjust = -1.5))

# Bottom plot
p2 <- base_plot +
  geom_point(data = filter(d_order, Location == "Bottom"), aes(x = x_coord, y = pvalue, color = Color), size = 0.75) +
  scale_y_reverse(name = expression(-log[10](italic(p)[int]) - " Downstream"), limits = c(15, 0), breaks = seq(0, 15, 5)) +
  theme(axis.text.x = element_blank(), axis.title.x = element_blank())

# Combine plots
combined_plot <- (p1 + guides(color = FALSE)) / (p2 + guides(color = FALSE))
print(combined_plot)
```

```{r load-gene-results}
gene_summary_df <- readRDS("../data/processed/manuscript/gene_summary_df.rds")
gc_summary_df <- readRDS("../data/processed/manuscript/gc_summary_df.rds")

gene_bonferroni <- 0.05 / length(unique(gene_summary_df$GENE))
```

```{r gwis-manhattan-genes}
# Make a Miami plot -- credit to RaMWAS package (pruning unneeded points)
# and hudson package (basis for Miami plot functionality)

# data <- filter(data, data[[pval_col_1]] > 0, data[[pval_col_2]] > 0)
nlps1 <- -log10(filter(gene_summary_df, pathway == "upstream")[["P"]])
nlps2 <- -log10(filter(gene_summary_df, pathway == "downstream")[["P"]])

# Trim points in crowded regions (credit to RaMWAS package for code snippet)
yfac = as.integer(nlps1 * 100) + 1L
yorder = sort.list(yfac)
levels(yfac) = as.character(seq_len(max(yfac)))
class(yfac) = "factor"
ygroup = split(seq_along(yfac), yfac)
for (i in seq_along(ygroup)) {
  if (length(ygroup[[i]]) > 300) {
    ygroup[[i]] = sample(ygroup[[i]], size=300, replace=FALSE)
  }
}
keep1 = unlist(ygroup, use.names=FALSE)
yfac = as.integer(nlps2 * 100) + 1L
yorder = sort.list(yfac)
levels(yfac) = as.character(seq_len(max(yfac)))
class(yfac) = "factor"
ygroup = split(seq_along(yfac), yfac)
for (i in seq_along(ygroup)) {
  if (length(ygroup[[i]]) > 300) {
    ygroup[[i]] = sample(ygroup[[i]], size=300, replace=FALSE)
  }
}
keep2 = unlist(ygroup, use.names=FALSE)

top <- cbind(
  select(filter(gene_summary_df, pathway == "upstream"), SYMBOL, CHR, POS = START), 
  pvalue=nlps1
)[keep1, ]
bottom=cbind(
  select(filter(gene_summary_df, pathway == "downstream"), SYMBOL, CHR, POS = START), 
  pvalue=nlps2
)[keep2, ]

topn <- names(top)
bottomn <- names(bottom)
top$Location <- "Top"
bottom$Location <- "Bottom"
d <- rbind(top, bottom)
d$POS <- as.numeric(as.character(d$POS))
d$CHR <- factor(d$CHR, levels = c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "X", "Y"))
d_order <- d[order(d$CHR, d$POS), ]
d_order$pos_index <- seq.int(nrow(d_order))
chr_lengths <- sapply(1:22, function(chr) max(d[d$CHR == chr, "POS"]))
chr_start_pos <- cumsum(chr_lengths) - chr_lengths
d_order$x_coord <- chr_start_pos[d_order$CHR] + d_order$POS
d_order_sub <- d_order[, c("SYMBOL", "CHR", "POS", "pvalue", "pos_index", "x_coord")]
maxRows <- by(d_order_sub, d_order_sub$CHR, function(x) x[which.max(x$x_coord),])
minRows <- by(d_order_sub, d_order_sub$CHR, function(x) x[which.min(x$x_coord),])
milimits <- do.call(rbind, minRows)
malimits <- do.call(rbind, maxRows)
lims <- merge(milimits, malimits, by="CHR")
names(lims) <- c("Color", 
                 "snpx", "posx", "px", "posidxx", "xcoordmin",
                 "snpy", "posy", "py", "posidxy", "xcoordmax")
lims$av <- (lims$xcoordmin + lims$xcoordmax)/2
lims <- lims[order(lims$Color),]

d_order$Color <- as.character(d_order$CHR)
newcols <-c(rep(x=c("#AAAAAA", "#8A8A8A"), length.out=22, each=1),  # Gray/dark gray for alternating chromosomes
            RColorBrewer::brewer.pal(3, "Dark2")[c(2, 3, 1)])
names(newcols) <-c(levels(factor(lims$Color)), "30", "31", "32")
d_order <- arrange(d_order, as.integer(Color), desc(pvalue)) %>%
  distinct(SYMBOL, Location, .keep_all=T)

# TOP PLOT
p1 <- ggplot() +
  geom_point(data=d_order[d_order$Location=="Top", ], 
             aes(x=x_coord, y=pvalue, color=factor(Color)), 
             size=0.75, alpha=1) +
  geom_hline(yintercept=-log10(gene_bonferroni), linetype="dashed", color="black") + 
  scale_x_continuous(breaks=lims$av[c(1:16, 18, 20, 20, 22)], 
                     labels=lims$Color[c(1:16, 18, 20, 20, 22)], 
                     expand=c(0,0)) +
  scale_y_continuous(
    limits=c(0, 15), breaks = seq(0, 15, 5), expand = c(0, 0),
    name=expression(-log[10](italic(p)) - Upstream)
  ) +
  scale_colour_manual(name = "Color", values = newcols) +
  scale_fill_manual(name = "Color", values = newcols) +
  theme(axis.title.x=element_blank(),
        axis.text.x = element_text(vjust = -1.5),
        panel.grid.minor = element_blank(),
        panel.grid.major.x = element_blank())

# BOTTOM PLOT
p2 <- ggplot(data=filter(d_order, Location == "Bottom"),
             aes(x=x_coord, y=pvalue, color=factor(Color))) +
  geom_point(size=0.75, alpha=1) +
  geom_hline(yintercept=-log10(gene_bonferroni), linetype="dashed", color="black") + 
  scale_x_continuous(breaks=lims$av[c(1:16, 18, 20, 20, 22)], 
                     labels=lims$Color[c(1:16, 18, 20, 20, 22)], 
                     expand=c(0,0)) +
  scale_colour_manual(name = "Color", values = newcols,
                      # breaks=c(30, 31),
                      # labels=c("Genome-wide significant joint test, no genome-wide significant marginal test",
                      #          "Genome-wide significant interaction test")
  ) +
  scale_fill_manual(name = "Color", values = newcols) +
  ylab(expression(-log[10](p) - Interaction)) +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank(),
        panel.grid.minor = element_blank(),
        panel.grid.major.x = element_blank()) +
  scale_y_reverse(
    limits=c(15,0), breaks = seq(0, 10, 5), expand = c(0, 0),
    name=expression(-log[10](italic(p)) - Downstream)
  )

(p1 + guides(color = "none")) /  
  (p2 + guides(color = "none"))
```

```{r gene-summary-plot}
make_gene_summary_manhattan <- function(data, 
                                        pval_col = "P", snp_col = "SYMBOL", 
                                        chr_col = "CHR", pos_col = "START", 
                                        threshold = gene_bonferroni, 
                                        ylims = NULL, main = "",
                                        add_labels = TRUE) {
  
  mh_data <- data %>%
    dplyr::rename(all_of(c(SNP = snp_col, CHR = chr_col, POS = pos_col, P = pval_col))) %>%
    filter(!is.na(P)) %>%
    mutate(P = as.numeric(P),
           P = ifelse(P == 0, min(1e-300, min(P[P != 0], na.rm = TRUE)), P),  # Remove P = 0
           nlp = -log10(P))
  
  # Trim points in crowded regions (credit to RaMWAS package for code snippet)
  yfac = as.integer(mh_data$nlp * 100) + 1L
  yorder = sort.list(yfac)
  yfac <- factor(yfac, levels = as.character(seq_len(max(yfac))))
  ygroup <- split(seq_along(yfac), yfac)
  for (i in seq_along(ygroup)) {
    if (length(ygroup[[i]]) > 300) {
      ygroup[[i]] <- sample(ygroup[[i]], size = 300, replace = FALSE)
    }
  }
  keep <- unlist(ygroup, use.names = FALSE)
  
  mh_data <- mh_data %>%
    select(SNP, CHR, POS, nlp, pathway, fdr_sig) %>%
    dplyr::slice(keep) %>%
    mutate(POS = as.numeric(as.character(POS)),
           CHR = factor(CHR, levels = c(1:22, "X"))) %>%
    arrange(CHR, POS) %>%
    mutate(pos_idx = seq(1, nrow(.)))
  
  suppressWarnings(chr_lengths <- sapply(c(1:22, "X"), function(chr) {
    with(mh_data, max(POS[CHR == chr], na.rm = TRUE))
  }))
  chr_lengths <- ifelse(is.infinite(chr_lengths), 0, chr_lengths)
  chr_start_pos <- cumsum(chr_lengths) - chr_lengths
  
  mh_data <- mh_data %>%
    mutate(x_coord = chr_start_pos[CHR] + POS,
           color = ifelse(fdr_sig, pathway, CHR),
    )
  
  lims <- mh_data %>%
    group_by(CHR) %>%
    summarise(avg_coord = (min(x_coord) + max(x_coord)) / 2)
  
  newcols <- setNames(
    c(rep(x = c("#AAAAAA", "#8A8A8A"), length.out = 22),
      RColorBrewer::brewer.pal(3, "Dark2")),
    c(levels(factor(lims$CHR)), unique(mh_data$pathway))
  )
  
  mh_plt <- ggplot() +
    geom_point(data = mh_data, 
               aes(x = x_coord, y = nlp, color = factor(color)), 
               size = 0.75, alpha = 1) +
    geom_hline(yintercept = -log10(threshold), linetype = "dashed", color = "gray") + 
    scale_x_continuous(breaks = lims$avg_coord[c(1:16, 18, 20, 20, 22, 23)], 
                       labels = c(1:16, 18, 20, 20, 22, "X"), 
                       expand = c(0,0)) +
    scale_y_continuous(name = expression(-log[10](italic(p)[int]))) +
    scale_colour_manual(values = newcols, breaks = unique(mh_data$pathway),
                        name = "Sub-pathway") +
    scale_fill_manual(name = "Color", values = newcols) +
    labs(title = main) +
    theme(axis.title.x = element_blank(),
          axis.text.x = element_text(vjust = -1.5),
          panel.grid.minor = element_blank(),
          panel.grid.major.x = element_blank())
  if (!is.null(ylims)) mh_plt <- mh_plt + coord_cartesian(ylim = ylims)
  if (add_labels) mh_plt <- mh_plt +
    ggrepel::geom_text_repel(data = filter(mh_data, fdr_sig),
                             aes(x = x_coord, y = nlp, label = SNP))
  
  mh_plt
}

gene_bonferroni <- 0.05 / length(unique(gene_summary_df$GENE))

make_gene_summary_manhattan(gene_summary_df)
```

Bonferroni significance threshold for the above plot is `r gene_bonferroni`, based on correction for `r length(unique(gene_summary_df$GENE))` unique genes.

```{r gene-summary-table}
gene_summary_tbl <- gene_summary_df %>%
  mutate(locus = paste0(str_pad(CHR, 2, pad = "0"), ":", START, "-", STOP)) %>%
  arrange(P) %>%
  filter(fdr_sig) %>%
  select(Pathway = pathway, Gene = SYMBOL, Locus = locus, `# SNPs` = NSNPS, 
         `Gene P` = P, `FDR q` = q)
# gene_summary_tbl %>%
#   write_csv("../data/processed/magma/top_genes.csv")
gene_summary_tbl %>%
  kable(caption = "Top gene-level findings") %>%
  kable_styling(full_width = FALSE)
```

```{r genetic-correlations}
# gc_summary_df %>%
#   mutate(across(c(p1, p2), ~ gsub("\\.sumstats\\.gz", "", basename(.x))),
#          y = ifelse(grepl("hscrp_log$", p1), "hscrp_log", "Omega_3_pct"),
#          l95 = rg - 1.96 * se,
#          u95 = rg + 1.96 * se) %>%
#   ggplot(aes(x = y, y = rg)) +
#   geom_errorbar(aes(ymin = l95, ymax = u95), width = 0.2) +
#   geom_point() +
#   # geom_bar(stat = "identity") +
#   geom_hline(yintercept = c(0, 1), color = "gray") +
#   coord_cartesian(ylim = c(0, 2)) +
#   labs(x = "Outcome trait", y = "GxE rho_g: oily fish & fish oil")

gc_summary_df %>%
  mutate(across(c(p1, p2), ~ gsub("\\.sumstats\\.gz", "", basename(.x))),
         y = ifelse(grepl("hscrp_log$", p1), "hscrp_log", "Omega_3_pct"),
         p = round(p, 2)) %>%
  ggplot(aes(x = y, y = rg)) +
  geom_bar(stat = "identity", width = 0.8) +
  geom_text(aes(label = paste0("p = ", p)), nudge_y = -0.2, color = "white") +
  geom_hline(yintercept = c(0, 1), color = "gray") +
  labs(x = "Outcome trait", y = expression(rho[g] * ": G x Fish and G x Fish Oil"))
```

# Sensitivity analysis

```{r sensitivity-setup}
run_gxe <- function(g, e, y, 
                    covars = c("age", "sex"), 
                    df = phenos,
                    std = FALSE, robust = FALSE) {
  if (std) df <- mutate(df, across(all_of(c(e, y)), ~ scale(.)))
  df$g <- df[[g]]
  lm_str <- paste0(y, " ~ g * ", e, " + ", paste0(covars, collapse = " + "))
  lm_summ <- tryCatch({
    lm_fit <- lm(as.formula(lm_str), data = df) 
    residual_dof <- lm_fit$df.residual
    if (robust) {
      lm_fit <- lmtest::coeftest(lm_fit, vcov = sandwich::vcovHC, type = "HC0")
    }
    lm_fit %>%
      broom::tidy() %>%
      filter(term == paste0("g:", e)) %>%
      mutate(residual_df = residual_dof)
  }, error = tibble(NA))
  lm_summ
}

run_sensitivity <- function(gene, e, y, 
                            base_covars = covar_sets$ffqAdj_minusAC,
                            alternate_e = NULL, alternate_y = NULL,
                            snp_idx = 1) {
  
  top_snp_df <- read_tsv(paste0("../data/processed/gene_followup/", 
                                e, "_", y, "_", gene, "_top_snp.raw"),
                         show_col_types = FALSE) %>%
    select(id = IID, last_col()) %>%
    mutate(id = as.character(id))
  top_snp <- names(top_snp_df)[snp_idx + 1]
  regression_df <- inner_join(phenos, top_snp_df, by = "id")
  
  e_regression <- if (e == "fish_oil") "fish_oil_touchscreen" else e
  if (!is.null(alternate_e)) e_regression <- alternate_e
  if (!is.null(alternate_y)) y <- alternate_y
  
  sensitivity_covar_list <- list(
    primary = base_covars,
    add_E_by_gPC = c(base_covars, paste0(e_regression, " * gPC", 1:10)),
    add_G_by_Cov = c(base_covars, paste0("g * ", base_covars)),
    add_BMI = c(base_covars, "bmi")
  )
  sensitivity_covar_res_df <- tibble(covar_set = names(sensitivity_covar_list)) %>%
    rowwise() %>%
    mutate(model_res = list(run_gxe(top_snp, e_regression, y,
                                    sensitivity_covar_list[[covar_set]],
                                    regression_df, std = TRUE))) %>%
    unnest(model_res) %>%
    mutate(across(estimate:p.value, ~ signif(., 3))) %>%
    mutate(snp = top_snp)
  sensitivity_covar_res_df
}
```

```{r sensitivity}
upstream_sensitivity_res <- tibble(
  gene = "myrf"
) %>%
  rowwise() %>%
  mutate(sensitivity_res = list(run_sensitivity(gene, "dn3fa", "Omega_3_pct"))) %>%
  unnest(sensitivity_res)
  
downstream_sensitivity_res <- tibble(
  gene = c("cbll1", "mica")
) %>%
  rowwise() %>%
  mutate(sensitivity_res = list(run_sensitivity(gene, "Omega_3_pct", "hscrp_log"))) %>%
  unnest(sensitivity_res)

sensitivity_res_tbl <- bind_rows(list(
  upstream = upstream_sensitivity_res, 
  downstream = downstream_sensitivity_res
), .id = "pathway") %>%
  mutate(gene = toupper(gene),
         result = paste0(round(estimate, 3), " (", format(p.value, scientific = TRUE), ")")) %>%
  select(gene, snp, pathway, covar_set, result) %>%
  pivot_wider(names_from = "covar_set", values_from = "result") %>%
  select(Gene = gene, Pathway = pathway, Variant = snp,
         Primary = primary,
         `Add E x gPC` = add_E_by_gPC, 
         `Add G x Covariate` = add_G_by_Cov, 
         `Add BMI` = add_BMI)

write_csv(sensitivity_res_tbl, "../output/manuscript/gene_sensitivity_res_df.csv")

sensitivity_res_tbl
```

# Replication

```{r wghs-replication}
wghs_rep_df %>%
  mutate(label = paste0(SYMBOL, " (", rsid, ")"),
         sig = p_wghs < 0.05) %>%
  filter(y == "LNCRP") %>%
  ggplot(aes(x = beta_ukb, y = beta_wghs, color = sig)) +
  geom_point() +
  geom_abline(slope = 1, intercept = 0)
```

# Vignettes

```{r prep-vignettes}
run_gxe <- function(g, e, y, 
                    covars = c("age", "sex"), 
                    df = phenos,
                    std = FALSE, robust = FALSE) {
  if (std) df <- mutate(df, across(all_of(c(e, y)), ~ scale(.)))
  df$g <- df[[g]]
  lm_str <- paste0(y, " ~ g * ", e, " + ", paste0(covars, collapse = " + "))
  lm_summ <- tryCatch({
    lm_fit <- lm(as.formula(lm_str), data = df) 
    residual_dof <- lm_fit$df.residual
    if (robust) {
      lm_fit <- lmtest::coeftest(lm_fit, vcov = sandwich::vcovHC, type = "HC0")
    }
    lm_fit %>%
      broom::tidy() %>%
      filter(term == paste0("g:", e)) %>%
      mutate(residual_df = residual_dof)
  }, error = tibble(NA))
  lm_summ
}

base_covars <- covar_sets$ffqAdj

phenos$oily_fish_fct <- factor(
  phenos$oily_fish, 
  labels = c("Never", "Less than once a week", "Once a week", 
             "2-4 times a week", "5-6 times a week", "Once or more daily")
)

format_p <- function(x) {
  ifelse(
    x > 0.001,
    round(x, 2),
    format(x, scientific = TRUE, digits = 2)
  )
}
```

## *FADS1* story

```{r collect-fads1-data}
gwis_sumstats_df <- read_tsv(paste0(
  "../data/processed/gene_followup/oily_fish_Omega_3_pct_fads1_subset"
))

top_snp_df <- read_tsv(paste0("../data/processed/gene_followup/oily_fish_Omega_3_pct_fads1_top_snp.raw")) %>%
  select(id = IID, last_col()) %>%
  mutate(id = as.character(id))
top_snp <- names(top_snp_df)[2]
gene_fu_df <- phenos %>%
  inner_join(top_snp_df, by = "id") %>%
  mutate(G = !! sym(top_snp))

all_maf_uncond_sumstats_df <- read_tsv("../data/processed/gene_followup/oily_fish_Omega_3_pct_fads1_regressions")
all_maf_cond_sumstats_df <- read_tsv("../data/processed/gene_followup/oily_fish_Omega_3_pct_fads1_regressions_cond")
variant_annot_df <- read_tsv(
  paste0("../data/processed/annovar/fads1_variants.variant_function"),
  col_names = c("annot", "gene", "chr", "start", "end", "REF", "ALT", "SNPID")
) %>%
  select(SNPID, annot)
all_maf_sumstats_df <- bind_rows(list(
  unconditional = all_maf_uncond_sumstats_df, 
  conditional = all_maf_cond_sumstats_df
), .id = "adjustment") %>%
  # filter(SE_Beta_G < (100 * median(SE_Beta_G))) %>%
  mutate(int_estimate = `Beta_G-oily_fish`,
         int_se = `SE_Beta_G-oily_fish`, 
         l95 = int_estimate - 1.96 * int_se,
         u95 = int_estimate + 1.96 * int_se,
         top_gwis = (SNPID == gsub("_.*", "", top_snp)),
         MAF = pmin(AF, 1 - AF)) %>%
  left_join(variant_annot_df, by = "SNPID") %>%
  mutate(annot = ifelse(is.na(annot), "unannotated", annot))
```

```{r fads1-pathways}
gene_summary_df %>%
  filter(grepl("FADS", SYMBOL)) %>%
  select(pathway, SYMBOL, P) %>%
  mutate(P = format_p(P)) %>%
  pivot_wider(names_from = "SYMBOL", values_from = "P") %>%
  kable(caption = "Decomposed gene-level significance for FADS genes") %>%
  kable_styling(full_width = FALSE)
```

```{r fads1-sensitivity}
sensitivity_covar_list <- list(
  primary = base_covars,
  add_E_by_gPC = c(base_covars, paste0("oily_fish * gPC", 1:10)),
  add_G_by_Cov = paste0("g * ", base_covars),
  add_BMI = c(base_covars, "bmi")
)
sensitivity_covar_res_df <- tibble(covar_set = names(sensitivity_covar_list)) %>%
  rowwise() %>%
  mutate(model_res = list(run_gxe(top_snp, "oily_fish", "Omega_3_pct",
                                  sensitivity_covar_list[[covar_set]], 
                                  gene_fu_df, std = TRUE))) %>%
  unnest(model_res) %>%
  mutate(across(estimate:p.value, ~ signif(., 3))) %>%
  select(`Covariate set` = covar_set, `GxE estimate` = estimate, P = p.value)
kable(sensitivity_covar_res_df,
      caption = "Sensitivity analyses: choice of covariates") %>%
  kable_styling(full_width = FALSE)

sensitivity_e_list <- list(
  `Oily fish` = "oily_fish",
  `Non-oily fish` = "nonoily_fish",
  `Total N3FA from 24HR` = "N3FA",
  `Fish oil supplementation` = "fish_oil_touchscreen"
)
sensitivity_e_res_df <- tibble(e_test = names(sensitivity_e_list)) %>%
  rowwise() %>%
  mutate(model_res = list(run_gxe(top_snp, sensitivity_e_list[[e_test]], 
                                  "Omega_3_pct",
                                  base_covars, 
                                  gene_fu_df, std = TRUE))) %>%
  unnest(model_res) %>%
  mutate(across(estimate:p.value, ~ signif(., 3))) %>%
  select(`Exposure` = e_test, `GxE estimate` = estimate, P = p.value)
kable(sensitivity_e_res_df,
      caption = "Sensitivity analyses: alternate exposures") %>%
  kable_styling(full_width = FALSE)
```

```{r fads1-viz}
gene_fu_df %>%
  mutate(G = factor(round(G))) %>%
  filter(!is.na(G), !is.na(oily_fish), !is.na(Omega_3_pct)) %>%
  group_by(G, oily_fish_fct) %>%
  summarise(m = mean(Omega_3_pct),
            se = sd(Omega_3_pct) / sqrt(n())) %>%
  mutate(l95 = m - 1.96 * se,
         u95 = m + 1.96 * se) %>%
  ggplot(aes(x = G, y = m, color = oily_fish_fct)) +
  geom_point(position = position_dodge(width = 0.2)) +
  geom_errorbar(aes(ymin = l95, ymax = u95), width = 0.2,
                position = position_dodge(width = 0.2)) +
  scale_color_discrete(name = "Oily fish intake") +
  labs(x = "Genotype", y = "Mean plasma N3FA% (95% CI)")

# beta_vec <- all_maf_sumstats_df[["Beta_G-oily_fish"]]
# ylims <- median(beta_vec, na.rm = TRUE) + c(-5, 5) * mad(beta_vec, na.rm = TRUE)

int_effect_maf_plt <- all_maf_sumstats_df %>%
  filter(adjustment == "unconditional") %>%
  arrange(top_gwis) %>%
  ggplot(aes(x = MAF, y = int_estimate)) +
  geom_point(aes(color = top_gwis)) +
  geom_errorbar(aes(color = top_gwis, ymin=l95, ymax=u95), width = 0) +
  geom_hline(yintercept = 0, color="gray") +
  scale_x_log10() +
  scale_color_manual(values = c(`FALSE` = "black", `TRUE` = "red"), 
                     breaks = c(TRUE),
                     labels = c("Top variant from GWIS")) +
  labs(x = "MAF", y = "Interaction effect estimate (95% CI)") +
  theme(axis.text.x = element_text(angle = 45, hjust = 0.9),
        legend.position = "bottom", legend.title = element_blank(), 
        legend.key = element_blank())

int_effect_cond_plt <- all_maf_sumstats_df %>%
  group_by(SNPID) %>%
  filter(any(P_Value_Interaction < 0.05)) %>%
  ungroup() %>%
  mutate(int_estimate = `Beta_G-oily_fish`,
         int_se = `SE_Beta_G-oily_fish`) %>%
  select(SNPID, AF, adjustment, int_estimate, int_se) %>%
  arrange(int_estimate) %>%
  mutate(l95 = int_estimate - 1.96 * int_se,
         u95 = int_estimate + 1.96 * int_se,
         top_gwis = (SNPID == gsub("_.*", "", top_snp)),
         MAF = pmin(AF, 1 - AF),
         SNPID = factor(SNPID, levels = unique(SNPID[adjustment == "unconditional"]))) %>%
  filter(!top_gwis) %>%
  ggplot(aes(x = SNPID, y = int_estimate, color = adjustment)) +
  geom_point(position = position_dodge(width = 0.1)) +
  geom_errorbar(aes(ymin = l95, ymax = u95), width = 0, 
                position = position_dodge(width = 0.1)) +
  geom_hline(yintercept = 0, color = "gray") + 
  labs(x = "", 
       y = "Interaction effect (95% CI)") +
  # coord_cartesian(ylim = ylims) +
  theme(axis.text.x = element_text(angle = 45, hjust = 0.9),
        legend.position = "bottom", legend.title = element_blank(), 
        legend.key = element_blank())

int_p_cond_plt <- all_maf_sumstats_df %>%
  mutate(int_nlp = -log10(P_Value_Interaction)) %>%
  select(SNPID, SNPID, MAF, adjustment, int_nlp, top_gwis) %>%
  pivot_wider(names_from = "adjustment", values_from = "int_nlp") %>%
  filter(!top_gwis) %>%
  ggplot(aes(x = unconditional, y = conditional, color = MAF)) +
  geom_point() +
  geom_abline(slope = 1, intercept = 0, color="gray") +
  scale_color_continuous(trans = "log", breaks = c(0.001, 0.01, 0.1)) +
  labs(x = "Interaction p-value (unconditional)", 
       y = "Interaction p-value (conditional on top variant)") +
  theme(axis.text.x = element_text(angle = 45, hjust = 0.9),
        legend.position = "bottom")

print(int_effect_maf_plt)
print(int_effect_cond_plt)
print(int_p_cond_plt)
```

```{r fads1-diet-corr}
for (e in c("dn3fa", unlist(sensitivity_e_list))) {
  gene_fu_df$G <- gene_fu_df[[top_snp]]
  g_e_lm_str <- paste0(e, " ~ G")
  g_e_lm_res_df <- lm(as.formula(g_e_lm_str), data = gene_fu_df) %>%
    broom::tidy()
  print(paste0(e, " ~ G: (p = ", round(g_e_lm_res_df$p.value[2], 3), ")"))
}
```

```{r hscrp-var-decomp}
# hscrp_covar_lm_str <- paste0("hscrp_log ~ ", paste(base_covars, collapse = " + "))
# hscrp_covar_lm <- lm(as.formula(hscrp_covar_lm_str), data = gene_fu_df)
# summary(hscrp_covar_lm)$r.squared

pn3fa_oily_fish_lm_str <- paste0("Omega_3_pct ~ oily_fish")
pn3fa_oily_fish_lm <- lm(as.formula(pn3fa_oily_fish_lm_str), data = gene_fu_df)
print("oily fish -> pN3FA R-squared:")
summary(pn3fa_oily_fish_lm)$r.squared

hscrp_pn3fa_lm_str <- paste0("hscrp_log ~ Omega_3_pct")
hscrp_pn3fa_lm <- lm(as.formula(hscrp_pn3fa_lm_str), data = gene_fu_df)
print("pN3FA -> log(hsCRP) R-squared:")
summary(hscrp_pn3fa_lm)$r.squared
```

## Main effects

### FADS1

```{r fads1-main-effects, eval=F}
fads1_cv_df <- all_maf_sumstats_df %>%
  filter(adjustment == "unconditional",
         MAF >= 0.01) %>%
  fads1_cv_marg_gxe_corr <- with(fads1_cv_df, cor(Beta_Marginal, `Beta_G-oily_fish`))

fads1_cv_df %>%
  ggplot(aes(x = Beta_Marginal, y = `Beta_G-oily_fish`)) +
  geom_point() + 
  geom_smooth(method = "lm", se = FALSE) +
  labs(title = paste0("Pearson correlation between marginal and GxE effects: ", 
                      round(fads1_cv_marg_gxe_corr, 2)))

# fads1_rv_df %>%
#   filter(P_Value_Marginal < 0.05,
#          annot != "ncRNA_exonic",
#          annot != "intronic") %>%
#   ggplot(aes(x = Beta_Marginal, y = `Beta_G-oily_fish`,
#              color = annot, group = annot)) +
#   geom_point() + 
#   geom_smooth(method = "lm", se = FALSE)

# fads1_rv_df %>%
#   filter(P_Value_Marginal < 0.05, #P_Value_Interaction < 0.05,
#          annot != "ncRNA_exonic") %>%
#   mutate(statistic_marg = Beta_Marginal / SE_Beta_Marginal,
#          statistic_gxe = `Beta_G-oily_fish` / `SE_Beta_G-oily_fish`) %>%
#   ggplot(aes(x = statistic_marg, y = `statistic_gxe`, color = MAF)) + 
#              # color = annot, group = annot)) +
#   geom_point() + 
#   geom_smooth(method = "lm", se = FALSE)

fads1_rv_df <- all_maf_sumstats_df %>%
  filter(adjustment == "conditional",
         MAF < 0.01, MAF > 0.001)

fads1_rv_df %>%
  mutate(l95 = Beta_Marginal - 1.96 * SE_Beta_Marginal,
         u95 = Beta_Marginal + 1.96 * SE_Beta_Marginal) %>%
  ggplot(aes(x = MAF, y = Beta_Marginal)) +
  geom_point() +
  geom_errorbar(aes(ymin = l95, ymax= u95), width = 0) +
  geom_hline(yintercept = 0, color = "gray") +
  scale_x_log10()

# fads1_rv_df %>%
#   mutate(beta_sign = factor(sign(Beta_Marginal))) %>%
#   ggplot(aes(x = annot, y = Beta_Marginal, fill = beta_sign)) +
#   geom_boxplot()
```

### FADS2

```{r fads2-main-effects, eval=F}
all_maf_uncond_sumstats_df <- read_tsv("../data/processed/gene_followup/oily_fish_Omega_3_pct_fads2_regressions")
all_maf_cond_sumstats_df <- read_tsv("../data/processed/gene_followup/oily_fish_Omega_3_pct_fads2_regressions_cond")
variant_annot_df <- read_tsv(
  paste0("../data/processed/annovar/fads2_variants.variant_function"),
  col_names = c("annot", "gene", "chr", "start", "end", "REF", "ALT", "SNPID")
) %>%
  select(SNPID, annot)
all_maf_sumstats_df <- bind_rows(list(
  unconditional = all_maf_uncond_sumstats_df, 
  conditional = all_maf_cond_sumstats_df
), .id = "adjustment") %>%
  filter(SE_Beta_G < (100 * median(SE_Beta_G))) %>%
  mutate(int_estimate = `Beta_G-oily_fish`,
         int_se = `SE_Beta_G-oily_fish`, 
         l95 = int_estimate - 1.96 * int_se,
         u95 = int_estimate + 1.96 * int_se,
         top_gwis = (SNPID == gsub("_.*", "", top_snp)),
         MAF = pmin(AF, 1 - AF)) %>%
  left_join(variant_annot_df, by = "SNPID") %>%
  mutate(annot = ifelse(is.na(annot), "unannotated", annot))

fads2_cv_df <- all_maf_sumstats_df %>%
  filter(adjustment == "unconditional",
         MAF >= 0.01)
fads2_cv_marg_gxe_corr <- with(fads2_cv_df, cor(Beta_Marginal, `Beta_G-oily_fish`))

fads2_cv_df %>%
  ggplot(aes(x = Beta_Marginal, y = `Beta_G-oily_fish`)) +
  geom_point() + 
  geom_smooth(method = "lm", se = FALSE) +
  labs(title = paste0("Pearson correlation between marginal and GxE effects: ", 
                      round(fads2_cv_marg_gxe_corr, 2)))

fads2_rv_df <- all_maf_sumstats_df %>%
  filter(adjustment == "conditional",
         MAF < 0.01, MAF > 0.001)

fads2_rv_df %>%
  mutate(l95 = Beta_Marginal - 1.96 * SE_Beta_Marginal,
         u95 = Beta_Marginal + 1.96 * SE_Beta_Marginal) %>%
  ggplot(aes(x = MAF, y = Beta_Marginal)) +
  geom_point() +
  geom_errorbar(aes(ymin = l95, ymax= u95), width = 0) +
  geom_hline(yintercept = 0, color = "gray") +
  scale_x_log10()

fads2_rv_df %>%
  filter(MAF > 0.001) %>%
  ggplot(aes(x = MAF, y = int_estimate)) +
  geom_point(aes(color = top_gwis)) +
  geom_errorbar(aes(color = top_gwis, ymin=l95, ymax=u95), width = 0) +
  geom_hline(yintercept = 0, color="gray") +
  scale_x_log10() +
  scale_color_manual(values = c(`FALSE` = "black", `TRUE` = "red"), 
                     breaks = c(TRUE),
                     labels = c("Top variant from GWIS")) +
  labs(x = "MAF", y = "Interaction effect estimate (95% CI)") +
  theme(axis.text.x = element_text(angle = 45, hjust = 0.9),
        legend.position = "bottom", legend.title = element_blank(), 
        legend.key = element_blank())

top_cond_snp_df <- read_tsv(paste0("../data/processed/gene_followup/oily_fish_Omega_3_pct_fads2_top_cond_snp.raw")) %>%
  select(id = IID, last_col()) %>%
  mutate(id = as.character(id))
top_cond_snp <- names(top_cond_snp_df)[2]
gene_fu_df <- phenos %>%
  inner_join(top_cond_snp_df, by = "id") %>%
  mutate(G = !! sym(top_cond_snp))
gene_fu_df %>%
  mutate(G = factor(round(G))) %>%
  filter(!is.na(G), !is.na(oily_fish), !is.na(Omega_3_pct)) %>%
  group_by(G, oily_fish_fct) %>%
  summarise(m = mean(Omega_3_pct),
            se = sd(Omega_3_pct) / sqrt(n())) %>%
  mutate(l95 = m - 1.96 * se,
         u95 = m + 1.96 * se) %>%
  ggplot(aes(x = G, y = m, color = oily_fish_fct)) +
  geom_point(position = position_dodge(width = 0.2)) +
  geom_errorbar(aes(ymin = l95, ymax = u95), width = 0.2,
                position = position_dodge(width = 0.2)) +
  scale_color_discrete(name = "Oily fish intake") +
  labs(x = "Genotype", y = "Mean plasma N3FA% (95% CI)")
```

```{r ggbio-testing, eval=F}
library(ggbio)
library(Homo.sapiens)

data(genesymbol,package = "biovizBase")
wh <- genesymbol[c("C11orf10", "FADS1", "FADS2", "FEN1", "C11orf9")]
wh <- range(wh, ignore.strand = TRUE)
fads_region_base_plt <- autoplot(Homo.sapiens, which = wh)
fads_region_base_plt +
  scale_x_continuous(name = "Position on chr11 (kb)", labels = ~ . / 1000)

wh <- genesymbol["FADS2"]
wh <- range(wh, ignore.strand = TRUE)
fads2_base_plt <- autoplot(Homo.sapiens, which = wh)

fads2_effects_plt <- fads2_rv_df %>%
  filter(MAF > 0.001) %>%
  ggplot(aes(x = POS, y = `Beta_G-oily_fish`)) +
  geom_point() +
  # geom_point(aes(color = top_gwis)) +
  geom_errorbar(aes(ymin=l95, ymax=u95), width = 0) +
  geom_hline(yintercept = 0, color="gray") +
  # scale_color_manual(values = c(`FALSE` = "black", `TRUE` = "red"), 
  #                    breaks = c(TRUE),
  #                    labels = c("Top variant from GWIS")) +
  labs(x = "MAF", y = "GxE estimate (95% CI)") +
  theme(axis.text.x = element_text(angle = 45, hjust = 0.9),
        legend.position = "bottom", legend.title = element_blank(), 
        legend.key = element_blank())

fads2_marg_effects_plt <- fads2_rv_df %>%
  mutate(
    # Beta_Marginal = abs(Beta_Marginal),
    l95 = Beta_Marginal - 1.96 * SE_Beta_Marginal,
    u95 = Beta_Marginal + 1.96 * SE_Beta_Marginal) %>%
  ggplot(aes(x = POS, y = Beta_Marginal)) +
  geom_point() +
  geom_errorbar(aes(ymin=l95, ymax=u95), width = 0) +
  geom_smooth() +
  geom_hline(yintercept = 0, color="gray") +
  # scale_color_manual(values = c(`FALSE` = "black", `TRUE` = "red"), 
  #                    breaks = c(TRUE),
  #                    labels = c("Top variant from GWIS")) +
  labs(x = "MAF", y = "G marginal estimate (95% CI)") +
  theme(axis.text.x = element_text(angle = 45, hjust = 0.9),
        legend.position = "bottom", legend.title = element_blank(), 
        legend.key = element_blank())

tks <- tracks(`Gene model` = fads2_base_plt, 
              # `FADS2 marginal effects` = fads2_marg_effects_plt,
              `FADS2 interaction effects` = fads2_effects_plt)
tks
```

## Power calculations

```{r power-calcs, cache=1}
library(ESPRESSO.GxE)

param_grid <- expand_grid(
  maf = c(0.1, 0.25),
  gxe_beta = c(0.01, 0.025, 0.05, 0.075, 0.1),
  N = c(200000),
  env_reliability = 1
)

simulation_params <- tibble(
  scenario.id = 1:nrow(param_grid),
  seed.val = 1,
  numsims = 100,
  numcases = 1000,  # These are just to prevent errors - shouldn't affect results
  numcontrols = 1000,  #
  numsubjects = param_grid$N,
  interaction.OR = 1,  #
  interaction.efkt = param_grid$gxe_beta,
  p.val = 5e-8,
  power = 0.8
)

pheno_params <- tibble(
  scenario.id = 1:nrow(param_grid),
  pheno.model = 1,
  disease.prev = 0.1,
  pheno.mean = 0, 
  pheno.sd = 1,
  pheno.sensitivity = 1,   # These are just to prevent errors - shouldn't affect results
  pheno.specificity = 1,   #
  pheno.reliability = 1
)

# Genotype parameters
geno_params <- tibble(
  scenario.id = 1:nrow(param_grid),
  geno.model = 1,
  MAF = param_grid$maf,
  geno.efkt = 0,
  geno.sensitivity = 1,
  geno.specificity = 1
)

# Environment parameters
env_params <- data.frame(
  scenario.id = 1:nrow(param_grid),
  env.model = 1,
  env.prevalence = 0.1,
  env.efkt = 0,
  env.mean = 0,
  env.sd = 1,
  env.low.lim = 0,
  env.up.lim = 1,
  env.reliability = param_grid$env_reliability
)

run_simulations <- function(sp, pp, gp, ep, s2r) {
  system("rm -f output.csv")  # Just to be sure (since ESPRESSO.GxE doesn't overwrite)
  capture.output(ESPRESSO.GxE::run.espresso.GxE(sp, pp, gp, ep, s2r))
  results <- suppressWarnings(read_delim("output.csv", delim = ";", 
                                         show_col_types = FALSE))
  new_colnames <- c(names(results)[1:34], 1:4, names(results)[35:38])
  results_fmt <- read_delim("output.csv", delim = ";", 
                            skip = 1, col_names = FALSE, show_col_types = FALSE) %>%
    setNames(new_colnames) %>%
    select(numsubjects, interaction.efkt, MAF, env.reliability,
           empirical.power)
  system("rm output.csv")
  results_fmt
}

sim_res <- run_simulations(simulation_params, pheno_params, geno_params, 
                           env_params, 1:nrow(param_grid))
```

```{r plot-power-calcs}
sim_res %>%
  mutate(MAF = factor(MAF)) %>%
  ggplot(aes(x = interaction.efkt, y = empirical.power, 
             group = MAF, color = MAF)) +
  geom_point() +
  geom_line() +
  # geom_vline(xintercept = 0.3, linetype = "dotted", color = "black") +
  # geom_vline(xintercept = 0.6, linetype = "dotted", color = "black") +
  labs(x = expression("Interaction effect (" * SD[y] * " / " * SD[e] * " / allele)"), 
       y = expression("Empirical power (" * italic(p) * " < 0.05)"))
```

```{r power-sims, cache=1}
N <- 200000
maf <- 0.25
n_sim <- 500
alpha <- 5e-8

targeted_sim_res_df <- expand_grid(
  pv_gxe_m = 0.025 ** 2, 
  pv_m_y = 0.009, 
  icc_m = 1
) %>%
  rowwise() %>%
  mutate(sim_res = list(test_upstream_scenario(n_sim, alpha, pv_gxe_m, pv_m_y, icc_m))) %>%
  unnest(sim_res) %>%
  mutate(pwr_se = sqrt(pwr * (1 - pwr) / n_sim),
         pwr_l95 = pwr - 1.96 * pwr_se,
         pwr_u95 = pwr + 1.96 * pwr_se)
```

```{r plot-power-sims}
targeted_sim_res_df

# targeted_sim_res_df %>%
#   mutate(across(c(pv_gxe_m), ~ factor(.x)),
#          approach = factor(approach,
#                            levels = c("std", "med"),
#                            labels = c("Standard", "Upstream"))) %>%
#   rename(`PV (GxE -> M)` = pv_gxe_m) %>%
#   ggplot(aes(x = pv_m_y, y = pwr, group = approach, color = approach)) +
#   geom_point() +
#   geom_line() +
#   scale_color_discrete(name = "Approach") + 
#   facet_wrap(vars(`PV (GxE -> M)`), ncol = 1, labeller = label_both) +
#   labs(x = "Proportion of variance in Y explained by M",
#        y = "Statistical power")
```

## *SLC26A4*

```{r collect-slc26a4-data, eval=F}
gwis_sumstats_df <- read_tsv(paste0(
  "../data/processed/gene_followup/Omega_3_pct_hscrp_log_slc26a4_subset"
))

top_snp_df <- read_tsv(paste0("../data/processed/gene_followup/Omega_3_pct_hscrp_log_slc26a4_top_snp.raw")) %>%
  select(id = IID, last_col()) %>%
  mutate(id = as.character(id))
top_snp <- names(top_snp_df)[2]
gene_fu_df <- phenos %>%
  inner_join(top_snp_df, by = "id") %>%
  mutate(G = !! sym(top_snp))

gene_fu_df %>%
  mutate(G = factor(round(G))) %>%
  filter(!is.na(G), !is.na(oily_fish), !is.na(Omega_3_pct)) %>%
  group_by(G, oily_fish) %>%
  summarise(m = mean(Omega_3_pct),
            se = sd(Omega_3_pct) / sqrt(n())) %>%
  mutate(l95 = m - 1.96 * se,
         u95 = m + 1.96 * se) %>%
  ggplot(aes(x = G, y = m, color = oily_fish)) +
  geom_point(position = position_dodge(width = 0.2)) +
  geom_errorbar(aes(ymin = l95, ymax = u95), width = 0.2,
                position = position_dodge(width = 0.2)) +
  labs(x = "Genotype", y = "Mean of Y (95% CI)")

sensitivity_covar_list <- list(
  primary = base_covars,
  add_E_by_gPC = c(base_covars, paste0("oily_fish * gPC", 1:10)),
  add_G_by_Cov = paste0("g * ", base_covars),
  add_BMI = c(base_covars, "bmi")
)
sensitivity_covar_res_df <- tibble(covar_set = names(sensitivity_covar_list)) %>%
  rowwise() %>%
  mutate(model_res = list(run_gxe(top_snp, "oily_fish", "Omega_3_pct",
                                  sensitivity_covar_list[[covar_set]], 
                                  gene_fu_df, std = TRUE))) %>%
  unnest(model_res) %>%
  mutate(across(estimate:p.value, ~ signif(., 3))) %>%
  select(`Covariate set` = covar_set, `GxE estimate` = estimate, P = p.value)
sensitivity_covar_res_df %>%
  mutate(P = format_p(P)) %>%
  kable(caption = "Sensitivity analyses: choice of covariates") %>%
  kable_styling(full_width = FALSE)

sensitivity_e_list <- list(
  `Oily fish` = "oily_fish",
  `Non-oily fish` = "nonoily_fish",
  `Total N3FA from 24HR` = "N3FA",
  `Fish oil supplementation` = "fish_oil_touchscreen"
)
sensitivity_e_res_df <- tibble(e_test = names(sensitivity_e_list)) %>%
  rowwise() %>%
  mutate(model_res = list(run_gxe(top_snp, sensitivity_e_list[[e_test]], 
                                  "Omega_3_pct",
                                  base_covars, 
                                  gene_fu_df, std = TRUE))) %>%
  unnest(model_res) %>%
  mutate(across(estimate:p.value, ~ signif(., 3))) %>%
  select(`Exposure` = e_test, `GxE estimate` = estimate, P = p.value)
sensitivity_e_res_df %>%
  mutate(P = format_p(P)) %>%
  kable(caption = "Sensitivity analyses: alternate exposures") %>%
  kable_styling(full_width = FALSE)

all_maf_uncond_sumstats_df <- read_tsv("../data/processed/gene_followup/Omega_3_pct_hscrp_log_slc26a4_regressions")
all_maf_cond_sumstats_df <- read_tsv("../data/processed/gene_followup/Omega_3_pct_hscrp_log_slc26a4_regressions_cond")
all_maf_sumstats_df <- bind_rows(list(
  unconditional = all_maf_uncond_sumstats_df, 
  conditional = all_maf_cond_sumstats_df
), .id = "adjustment") %>%
  filter(SE_Beta_G < (100 * median(SE_Beta_G))) %>%
  mutate(int_estimate = `Beta_G-oily_fish`,
         int_se = `SE_Beta_G-oily_fish`, 
         l95 = int_estimate - 1.96 * int_se,
         u95 = int_estimate + 1.96 * int_se,
         top_gwis = (SNPID == gsub("_.*", "", top_snp)),
         MAF = pmin(AF, 1 - AF))

beta_vec <- all_maf_sumstats_df[["Beta_G-oily_fish"]]
ylims <- median(beta_vec, na.rm = TRUE) + c(-5, 5) * mad(beta_vec, na.rm = TRUE)

int_effect_maf_plt <- all_maf_sumstats_df %>%
  filter(adjustment == "unconditional") %>%
  arrange(top_gwis) %>%
  ggplot(aes(x = MAF, y = int_estimate)) +
  geom_point(aes(color = top_gwis)) +
  geom_errorbar(aes(color = top_gwis, ymin=l95, ymax=u95), width = 0) +
  geom_hline(yintercept = 0, color="gray") +
  scale_x_log10() +
  scale_color_manual(values = c(`FALSE` = "black", `TRUE` = "red"), 
                     breaks = c(TRUE),
                     labels = c("Top variant from GWIS")) +
  labs(x = "MAF", y = "Interaction effect estimate (95% CI)") +
  theme(axis.text.x = element_text(angle = 45, hjust = 0.9),
        legend.position = "bottom", legend.title = element_blank(), 
        legend.key = element_blank())

int_effect_cond_plt <- all_maf_sumstats_df %>%
  group_by(SNPID) %>%
  filter(any(P_Value_Interaction < 0.05)) %>%
  ungroup() %>%
  mutate(int_estimate = `Beta_G-oily_fish`,
         int_se = `SE_Beta_G-oily_fish`) %>%
  select(SNPID, AF, adjustment, int_estimate, int_se) %>%
  arrange(int_estimate) %>%
  mutate(l95 = int_estimate - 1.96 * int_se,
         u95 = int_estimate + 1.96 * int_se,
         top_gwis = (SNPID == gsub("_.*", "", top_snp)),
         MAF = pmin(AF, 1 - AF),
         SNPID = factor(SNPID, levels = unique(SNPID[adjustment == "unconditional"]))) %>%
  filter(!top_gwis) %>%
  ggplot(aes(x = SNPID, y = int_estimate, color = adjustment)) +
  geom_point(position = position_dodge(width = 0.1)) +
  geom_errorbar(aes(ymin = l95, ymax = u95), width = 0, 
                position = position_dodge(width = 0.1)) +
  geom_hline(yintercept = 0, color = "gray") + 
  labs(x = "", 
       y = "Interaction effect (95% CI)") +
  # coord_cartesian(ylim = ylims) +
  theme(axis.text.x = element_text(angle = 45, hjust = 0.9),
        legend.position = "bottom", legend.title = element_blank(), 
        legend.key = element_blank())

int_p_cond_plt <- all_maf_sumstats_df %>%
  mutate(int_nlp = -log10(P_Value_Interaction)) %>%
  select(SNPID, SNPID, MAF, adjustment, int_nlp, top_gwis) %>%
  pivot_wider(names_from = "adjustment", values_from = "int_nlp") %>%
  filter(!top_gwis) %>%
  ggplot(aes(x = unconditional, y = conditional, color = MAF)) +
  geom_point() +
  geom_abline(slope = 1, intercept = 0, color="gray") +
  scale_color_continuous(trans = "log", breaks = c(0.001, 0.01, 0.1)) +
  labs(x = "Interaction p-value (unconditional)", 
       y = "Interaction p-value (conditional on top variant)") +
  theme(axis.text.x = element_text(angle = 45, hjust = 0.9),
        legend.position = "bottom")

print(int_effect_maf_plt)
print(int_effect_cond_plt)
print(int_p_cond_plt)

beta_int_col <- paste0("Beta_G-", e_regression)
top_cond_sumstats_df <- all_maf_cond_sumstats_df %>%
  filter(SE_Beta_G < (100 * median(SE_Beta_G))) %>%
  # arrange(desc(abs(!! sym(beta_int_col)))) %>%
  arrange(P_Value_Interaction) %>%
  select(SNPID, CHR, POS, Effect_Allele, 
         {{beta_int_col}}, P_Value_Interaction, P_Value_Marginal) %>%
  head()
print("Top variants in conditional analysis")
print(top_cond_sumstats_df)

top_cond_snp_df <- read_tsv(paste0("../data/processed/gene_followup/", 
                                   e, "_", y, "_", gene, "_top_cond_snp.raw")) %>%
  select(id = IID, last_col()) %>%
  mutate(id = as.character(id))
top_cond_snp <- names(top_cond_snp_df)[snp_idx + 1]
regression_df <- inner_join(phenos, top_cond_snp_df, by = "id")
cond_double_strat_plt <- make_double_strat_plot(top_cond_snp, e_regression, y, regression_df,
                                                "Stratified interaction plot for top conditional rare SNP")
cond_g_strat_plt <- make_g_strat_plot(top_cond_snp, e_regression, y, regression_df,
                                      "Stratified interaction plot for top conditional rare SNP")
print(cond_double_strat_plt + cond_g_strat_plt)
```


# AHA E|L outputs

```{r aha-1, fig.asp=1.2, fig.width=3, eval=F}
std_med_res_df <- readRDS("../data/processed/manuscript/std_med_res_df.rds") %>%
  mutate(tau.ci.lower = map_dbl(list(med_obj), \(x) x$tau.ci[1]),
         tau.ci.upper = map_dbl(list(med_obj), \(x) x$tau.ci[2]),
         d.ci.lower = map_dbl(list(med_obj), \(x) x$d.avg.ci[1]),
         d.ci.upper = map_dbl(list(med_obj), \(x) x$d.avg.ci[2]),
         z.ci.lower = map_dbl(list(med_obj), \(x) x$z.avg.ci[1]),
         z.ci.upper = map_dbl(list(med_obj), \(x) x$z.avg.ci[2]),
         d.coef = d.avg, z.coef = z.avg)

main_med_plt <- std_med_res_df %>%
  filter(e == "dn3fa", m == "Omega_3_pct", y == "hscrp_log") %>%
  select(contains("coef"), contains("ci")) %>%
  pivot_longer(everything(), names_to = c("estimate", "quantity"), 
               names_pattern = "([a-z]+)\\.(.*)", 
               values_to = "value") %>%
  pivot_wider(names_from = "quantity", values_from = "value") %>%
  mutate(estimate = factor(estimate, levels = c("tau", "d", "z"),
                           labels = c("Total effect", "Indirect (mediated) effect", "Direct effect"))) %>%
  ggplot(aes(x = estimate, y = coef)) +
  geom_point() +
  geom_errorbar(aes(ymin = ci.lower, ymax = ci.upper), width = 0) +
  geom_hline(yintercept = 0, color = "gray") +
  labs(y = "Estimate (dN3FA - pN3FA - hsCRP)") +
  theme(axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 20, hjust = 0.9))

main_med_plt
```

```{r aha-2, fig.asp=0.6, eval=F}
nlps1 <- -log10(upstream_gwis_df[["robust_P_int"]])
nlps2 <- -log10(downstream_gwis_df[["robust_P_int"]])

yfac = as.integer(nlps1 * 100) + 1L
yorder = sort.list(yfac)
levels(yfac) = as.character(seq_len(max(yfac)))
class(yfac) = "factor"
ygroup = split(seq_along(yfac), yfac)
for (i in seq_along(ygroup)) {
  if (length(ygroup[[i]]) > 300) {
    ygroup[[i]] = sample(ygroup[[i]], size=300, replace=FALSE)
  }
}
keep1 = unlist(ygroup, use.names=FALSE)
yfac = as.integer(nlps2 * 100) + 1L
yorder = sort.list(yfac)
levels(yfac) = as.character(seq_len(max(yfac)))
class(yfac) = "factor"
ygroup = split(seq_along(yfac), yfac)
for (i in seq_along(ygroup)) {
  if (length(ygroup[[i]]) > 300) {
    ygroup[[i]] = sample(ygroup[[i]], size=300, replace=FALSE)
  }
}
keep2 = unlist(ygroup, use.names=FALSE)

top <- cbind(select(upstream_gwis_df, SNP, CHR, POS), pvalue=nlps1)[keep1, ]
bottom=cbind(select(downstream_gwis_df, SNP, CHR, POS), pvalue=nlps2)[keep2, ]

topn <- names(top)
bottomn <- names(bottom)
top$Location <- "Top"
bottom$Location <- "Bottom"
d <- rbind(top, bottom)
d$POS <- as.numeric(as.character(d$POS))
d$CHR <- factor(d$CHR, levels = c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "X", "Y"))
d_order <- d[order(d$CHR, d$POS), ]
d_order$pos_index <- seq.int(nrow(d_order))
chr_lengths <- sapply(1:22, function(chr) max(d[d$CHR == chr, "POS"]))
chr_start_pos <- cumsum(chr_lengths) - chr_lengths
d_order$x_coord <- chr_start_pos[d_order$CHR] + d_order$POS
d_order_sub <- d_order[, c("SNP", "CHR", "POS", "pvalue", "pos_index", "x_coord")]
maxRows <- by(d_order_sub, d_order_sub$CHR, function(x) x[which.max(x$x_coord),])
minRows <- by(d_order_sub, d_order_sub$CHR, function(x) x[which.min(x$x_coord),])
milimits <- do.call(rbind, minRows)
malimits <- do.call(rbind, maxRows)
lims <- merge(milimits, malimits, by="CHR")
names(lims) <- c("Color", 
                 "snpx", "posx", "px", "posidxx", "xcoordmin",
                 "snpy", "posy", "py", "posidxy", "xcoordmax")
lims$av <- (lims$xcoordmin + lims$xcoordmax)/2
lims <- lims[order(lims$Color),]

# gw_int_snps <- whr_ss$RSID[whr_ss$P_Value_Interaction < 5e-8]
# gw_joint_noMarg_snps <- whr_ss$RSID[whr_ss$P_Value_Joint < 5e-8 &
#                                     whr_ss$P_Value_Main > 5e-8]
# d_order$Color <- case_when(
#   d_order$SNP %in% int_marg_snps ~ "30",
#   d_order$SNP %in% int_nomarg_snps ~ "31",
#   d_order$SNP %in% joint_nomarg_snps ~ "32",
#   TRUE ~ as.character(d_order$CHR)
# )
d_order$Color <- as.character(d_order$CHR)
newcols <-c(rep(x=c("#AAAAAA", "#8A8A8A"), length.out=22, each=1),  # Gray/dark gray for alternating chromosomes
            RColorBrewer::brewer.pal(3, "Dark2")[c(2, 3, 1)])
names(newcols) <-c(levels(factor(lims$Color)), "30", "31", "32")
d_order <- arrange(d_order, as.integer(Color), desc(pvalue)) %>%
  distinct(SNP, Location, .keep_all=T)

# TOP PLOT
p1 <- ggplot() +
  geom_point(data=d_order[d_order$Location=="Top", ], 
             aes(x=x_coord, y=pvalue, color=factor(Color)), 
             size=0.75, alpha=1) +
  geom_hline(yintercept=-log10(5e-8), linetype="dashed", color="black") + 
  scale_x_continuous(breaks=lims$av[c(1:16, 18, 20, 20, 22)], 
                     labels=lims$Color[c(1:16, 18, 20, 20, 22)], 
                     expand=c(0,0)) +
  scale_y_continuous(limits=c(0, 12), breaks = seq(0, 10, 5), expand = c(0, 0),
                     name=expression(-log[10](italic(p)) - Upstream)) +
  scale_colour_manual(name = "Color", values = newcols) +
  scale_fill_manual(name = "Color", values = newcols) +
  theme(axis.title.x=element_blank(),
        axis.text.x = element_text(vjust = -1.5),
        panel.grid.minor = element_blank(),
        panel.grid.major.x = element_blank())

# BOTTOM PLOT
p2 <- ggplot(data=filter(d_order, Location == "Bottom"),
             aes(x=x_coord, y=pvalue, color=factor(Color))) +
  geom_point(size=0.75, alpha=1) +
  geom_hline(yintercept=-log10(5e-8), linetype="dashed", color="black") + 
  scale_x_continuous(breaks=lims$av[c(1:16, 18, 20, 20, 22)], 
                     labels=lims$Color[c(1:16, 18, 20, 20, 22)], 
                     expand=c(0,0)) +
  scale_colour_manual(name = "Color", values = newcols,
                      # breaks=c(30, 31),
                      # labels=c("Genome-wide significant joint test, no genome-wide significant marginal test",
                      #          "Genome-wide significant interaction test")
  ) +
  scale_fill_manual(name = "Color", values = newcols) +
  ylab(expression(-log[10](p) - Interaction)) +
  theme(axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        panel.grid.minor = element_blank(),
        panel.grid.major.x = element_blank()) +
  scale_y_reverse(limits=c(12,0), breaks = seq(0, 10, 5), expand = c(0, 0),
                  name=expression(-log[10](italic(p)) - Downstream))

# color_legend <- get_legend(
#   p2 + 
#     theme(
#       legend.title=element_blank(), 
#       legend.background=element_rect(size=rel(0.5)),
#       legend.key.size=unit(0.1, "lines"),
#       legend.text=element_text(size=10)
#     ) + 
#     guides(color=guide_legend(override.aes=list(size=2.5)))
# )

(p1 + guides(color=F)) /  
  (p2 + guides(color=F))
```

```{r aha-3, fig.asp=1, fig.width=5, eval=F}
make_gene_summary_manhattan <- function(data, 
                                        pval_col = "P", snp_col = "SYMBOL", 
                                        chr_col = "CHR", pos_col = "START", 
                                        threshold = gene_bonferroni, 
                                        ylims = NULL, main = "",
                                        add_labels = TRUE) {
  
  mh_data <- data %>%
    dplyr::rename(all_of(c(SNP = snp_col, CHR = chr_col, POS = pos_col, P = pval_col))) %>%
    filter(!is.na(P)) %>%
    mutate(P = as.numeric(P),
           P = ifelse(P == 0, min(1e-300, min(P[P != 0], na.rm = TRUE)), P),  # Remove P = 0
           nlp = -log10(P))
  
  # Trim points in crowded regions (credit to RaMWAS package for code snippet)
  yfac = as.integer(mh_data$nlp * 100) + 1L
  yorder = sort.list(yfac)
  yfac <- factor(yfac, levels = as.character(seq_len(max(yfac))))
  ygroup <- split(seq_along(yfac), yfac)
  for (i in seq_along(ygroup)) {
    if (length(ygroup[[i]]) > 300) {
      ygroup[[i]] <- sample(ygroup[[i]], size = 300, replace = FALSE)
    }
  }
  keep <- unlist(ygroup, use.names = FALSE)
  
  mh_data <- mh_data %>%
    select(SNP, CHR, POS, nlp, pathway, fdr_sig) %>%
    dplyr::slice(keep) %>%
    mutate(POS = as.numeric(as.character(POS)),
           CHR = factor(CHR, levels = c(1:22, "X"))) %>%
    arrange(CHR, POS) %>%
    mutate(pos_idx = seq(1, nrow(.)))
  
  suppressWarnings(chr_lengths <- sapply(c(1:22, "X"), function(chr) {
    with(mh_data, max(POS[CHR == chr], na.rm = TRUE))
  }))
  chr_lengths <- ifelse(is.infinite(chr_lengths), 0, chr_lengths)
  chr_start_pos <- cumsum(chr_lengths) - chr_lengths
  
  mh_data <- mh_data %>%
    mutate(x_coord = chr_start_pos[CHR] + POS,
           color = ifelse(fdr_sig, pathway, CHR),
    )
  
  lims <- mh_data %>%
    group_by(CHR) %>%
    summarise(avg_coord = (min(x_coord) + max(x_coord)) / 2)
  
  newcols <- setNames(
    c(rep(x = c("#AAAAAA", "#8A8A8A"), length.out = 22),
      RColorBrewer::brewer.pal(3, "Dark2")),
    c(levels(factor(lims$CHR)), unique(mh_data$pathway))
  )
  
  mh_plt <- ggplot() +
    geom_point(data = mh_data, 
               aes(x = x_coord, y = nlp, color = factor(color)), 
               size = 0.75, alpha = 1) +
    geom_hline(yintercept = -log10(threshold), linetype = "dashed", color = "gray") + 
    scale_x_continuous(breaks = lims$avg_coord[c(1:16, 18, 20, 20, 22, 23)], 
                       labels = c(1:16, 18, 20, 20, 22, "X"), 
                       expand = c(0,0)) +
    scale_y_continuous(name = expression(-log[10](italic(p)[int]))) +
    scale_colour_manual(values = newcols, breaks = unique(mh_data$pathway),
                        name = "Sub-pathway") +
    scale_fill_manual(name = "Color", values = newcols) +
    labs(title = main) +
    theme(axis.title.x = element_blank(),
          axis.text.x = element_text(vjust = -1.5),
          panel.grid.minor = element_blank(),
          panel.grid.major.x = element_blank())
  if (!is.null(ylims)) mh_plt <- mh_plt + coord_cartesian(ylim = ylims)
  if (add_labels) mh_plt <- mh_plt +
    ggrepel::geom_text_repel(data = filter(mh_data, fdr_sig),
                             aes(x = x_coord, y = nlp, label = SNP))
  
  mh_plt
}

gene_bonferroni <- 0.05 / length(unique(gene_summary_df$GENE))

make_gene_summary_manhattan(gene_summary_df)
```